Proposal: Public Suffix API (#676)
hober/web-extensions: proposals/public-suffix.md

| example.co.uk         |
| example2.com          |
| foo.bar.example.co.uk |
| foo.bar.example2.com  |
| www.example.co.uk     |
| www.example2.com      |

##### Grouped domains

| example.co.uk  |                       |
| -------------- | --------------------- |
|                | example.co.uk         |
|                | foo.bar.example.co.uk |
|                | www.example.co.uk     |

| example2.com   |                       |
| -------------- | --------------------- |
|                | example2.com          |
|                | foo.bar.example2.com  |
|                | www.example2.com      |

### Known Consumers

Mozilla intends to make use of this API in its [multi-account-containers](https://github.com/mozilla/multi-account-containers) extension to allow users to create containers
that include wildcard domain names: [PR #2352](https://github.com/mozilla/multi-account-containers/pull/2352). Currently, users must manually curate their containers to capture all
anticipated subdomains of a base domain, and this manual approach is cumbersome and
error-prone.

Apart from this, any other extension that already rolls its own solution in order to parse

> the PSL could potentially benefit from this API, for example [uBlock Origin](https://github.com/gorhill/uBlock).

Reviewer: Could you be more specific in the specific use cases that depend on this proposed API? Knowing the use cases makes it easier to evaluate the proposed API.

Author: Agreed, the more use cases the better. I will take a look at the source code of other extensions to see how they are using the PSL data currently.

Author: Done 3301526fcb150706aa59e9f20839d646e4920649

## Specification

> ### Schema

Reviewer 2: Would you be open to adding a method to retrieve the PSL version in this proposal (I know it's mentioned in future work)? My understanding is that this currently doesn't exist, but the PSL maintainers have had requests for that in the past and were open to adding something. I'll let them know about this PR and perhaps they will have additional thoughts.

Author: Yes, happy to do so. Yes, it would be great if the PSL maintainers could let us know how their versioning will work.

Author: Done 3301526fcb150706aa59e9f20839d646e4920649

A new API `publicSuffix` is added as follows:

```ts
//
// Example:
//
// let domain = await browser.publicSuffix.getRegistrableDomain("www.example.co.uk");
// ==> 'example.co.uk'
//
Promise<string> browser.publicSuffix.getRegistrableDomain(
  // The domain name or URL whose registrable domain we want to find

>   domainNameOrUrl: string,

Reviewer: Is there any reason for accepting a domain or URL, instead of just one of them?

Author: No strong reason, happy to change to one or the other. Some extra info:
- There is a comment about this in the WECG minutes: ["Can take a domain or URL"](https://github.com/w3c/webextensions/blob/12d0eda63073e5b56205f03dad4b77aa8180d5ff/_minutes/2022-06-23-wecg.md?plain=1#L129)
- The `cookies` API has methods that take an object as an argument in which either a `url` or `domain` property can be set. We could consider doing the same here. See [cookies.getAll()](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/cookies/getAll)
- The `dns` API only accepts `hostname`. See [dns.resolve()](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/dns/resolve)
- The Chrome source code indicates that the version of `GetDomainAndRegistry` that accepts a URL parameter as opposed to a domain name string is [preferable from a performance perspective](https://source.chromium.org/chromium/chromium/src/+/main:net/base/registry_controlled_domains/registry_controlled_domain.h;l=191;drc=4f516be19f2c1d35dc3240d050d84d10f0d6f726). However, to benefit from this I assume the javascript parameter type would have to be `URL` as opposed to `string`, which may not be as ergonomic.

Author: Done - now accepts only `domain` 3301526fcb150706aa59e9f20839d646e4920649

)
```

### Behavior #1: Returned Promise

The promise returned by `getRegistrableDomain()` method will either:

1. Resolve with the topmost *registrable domain / eTLD+1* if it can be determined from
the `domainNameOrUrl`.

> 2. Reject with an `Error` / populate `browser.runtime.lastError`.

Reviewer: Could you list examples that are explicitly considered valid or error conditions? The way that this 2-step behavior is phrased is very light on details and the lack of detail makes it more difficult for people to know what the API does and what it does not.

Some quick examples of specific classes of inputs:
- `example.net` (eTLD+1)
- `www.example.net` (eTLD+2)
- `net` (this is a public suffix itself)
- `127.0.0.1` (IP address, IPv4)
- `[::1]` (IPv6 address)
- `.example.net` (dot in front)
- `example.net.` (dot in the end, this is a FQDN)
- `github.io` (PSL entry marked as PRIVATE)
- `*.com` (listed elsewhere in this proposal as a use case, but an invalid domain name)

Some more examples are at https://url.spec.whatwg.org/#example-host-psl . These examples include IDN, non-canonical domains (e.g. upper case).

Author: Ok I will add this.

Author: Done 3301526fcb150706aa59e9f20839d646e4920649

### Behavior #2: Private Domains

> By default, the lookup performed by `getRegistrableDomain()` should **exclude** private domains

Reviewer: Please add a reference to the meaning of private in this context.

Author: Ok will do

Author: Done 3301526fcb150706aa59e9f20839d646e4920649

contained in the PSL dataset.

See [Future work](#1-extend-the-api).

### New Permissions

| Permission Added | Suggested Warning |
| ---------------- | ----------------- |

> | publicSuffix     | Read the public suffix list |

Reviewer: The public suffix list is not private and I would therefore not associate a permission warning with it. There are already alternative ways to guess the public suffix for a domain, e.g. through the cookies API.

Author: Makes sense, I will remove the warning.

Author: Done 3301526fcb150706aa59e9f20839d646e4920649

## Implementation Notes

Since the major browsers all already implement internal methods for determining
topmost *registrable domain / eTLD+1*s, it is hoped that the implementation will
involve little more than providing the relevant mechanism for exposing these same methods
to extensions:

| Browser | Registrable Domain Method |
| ------- | ------------------------- |
| Chrome  | [GetDomainAndRegistry](https://source.chromium.org/chromium/chromium/src/+/main:net/base/registry_controlled_domains/registry_controlled_domain.h;l=182;drc=4f516be19f2c1d35dc3240d050d84d10f0d6f726) |
| Firefox | [getBaseDomain](https://searchfox.org/mozilla-central/rev/e9f9bf31d1c0057a1cd339b5a853a75d1b16db39/netwerk/dns/nsIEffectiveTLDService.idl#94) |
| Safari  | [topPrivatelyControlledDomain](https://github.com/WebKit/WebKit/blob/01eba7c416725cfd4eec57ab16daffa25b8124b4/Source/WebCore/platform/PublicSuffixStore.h#L43) |

## Future Work

### 1. Extend the API

The major browsers provide additional methods/parameters internally for getting
information related to the *registable domain / eTLD+1*. The API could be extended
to expose more of these internal methods/parameters, for example:

> 1. Provide an option to include private domains when getting the *registrable domain / eTLD+1*.

Reviewer 2: Could you say more about why this is considered future work? I think it would be quite important to have, and even wonder if it should be enabled by default. Are there common use cases where you wouldn't want to include the private section?

Author: I agree it may be better to have the private section included by default as you suggest, and am happy to make this change. Here are some points to consider:

#### 1. Transparency / simplicity
By excluding the private section, the API returns only domains registered with an ICANN registrar. With the private section included, the registry mechanism for the returned domain may be less clear.

##### Example

|  |  |
| - | - |
| Full Domain to lookup | *foo.bar.baz.execute-api.us-east-1.amazonaws.com* |
| Registrable Domain (excl. private) | *amazonaws.com* |
| Registrable Domain (incl. private) | *baz.execute-api.us-east-1.amazonaws.com* |

#### 2. Granularity
This API may be used to automatically divide up lists of URLs into groups by their *registrable domain*. If using the entries from the private section, you are likely to end up with a higher number of groups, and fewer URLs per group (i.e. more fine-grained grouping). For example:

##### Example: URLs to be Sorted Into Groups

| URLs |
| - |
| foo.execute-api.us-east-1.amazonaws.com/ |
| bar.s3.dualstack.us-east-1.amazonaws.com/ |
| baz.vfs.cloud9.us-east-1.amazonaws.com/ |

##### No Private Domains

| Registrable Domain (excl. private) | URLs |
| - | - |
| **amazonaws.com** | |
|   | foo.execute-api.us-east-1.amazonaws.com/       |
|   | bar.s3.dualstack.us-east-1.amazonaws.com/      |
|   | baz.vfs.cloud9.us-east-1.amazonaws.com/ |

##### Including Private Domains

| Registrable Domain (incl. private) | URLs |
| - | - |
| **foo.execute-api.us-east-1.amazonaws.com**    |  |
|  | foo.execute-api.us-east-1.amazonaws.com/ |
| **bar.s3.dualstack.us-east-1.amazonaws.com**   |  |
|   | bar.s3.dualstack.us-east-1.amazonaws.com/ |
| **baz.vfs.cloud9.us-east-1.amazonaws.com** |  |
|   | baz.vfs.cloud9.us-east-1.amazonaws.com/ |

Author: Done - now defaults to Private suffixes included 3301526fcb150706aa59e9f20839d646e4920649

**Related Issues:** [#231](https://github.com/w3c/webextensions/issues/231)

## Motivation

### Objective

This API enables developers to obtain the *registrable domain / eTLD+1*
from a domain name. This functionality is already implemented for internal
use by all the major browsers. Therefore the effect of this API is to expose
existing built-in browser functionality to extensions developers.

The primary objective of this API is to eliminate the possibility of inconsistencies
between the host browser and hosted extensions when deriving
*registrable domain / eTLD+1*s from domain names.

Secondary objectives of this API are to:

1. Improve extension developer experience by reducing complexity and maintenance overhead,
since developers will no longer need to roll their own solutions for obtaining and parsing
the [Public Suffix List (PSL)](https://publicsuffix.org/list/).
2. Reduce extensions' resource usage (CPU, memory, disk space), since extensions
will no longer be duplicating work already done by the host browser.

### Features of the PSL

In order to set out the use cases fully, it is first necessary to describe certain
relevant features of the PSL.

#### 1. ICANN vs Private

> The PSL it is divided into two sections: ICANN suffixes and Private (i.e. non-ICANN)

Reviewer 3: nit: delete "it"

Author: Sure

Author: Addressed in c2171564bebfbd15314e2e147a613129044a4753

suffixes.

##### Examples

| ICANN Suffixes                | Private Suffixes (Non-ICANN)  |
|-------------------------------|------------------------------ |
| com                           | blogspot.com                  |
|                               | members.linode.com            |
|                               | simplesite.com                |
|                               |                               |
| co.uk                         | blogspot.co.uk                |
|                               | wellbeingzone.co.uk           |
|                               | vm.bytemark.co.uk             |

The suffixes in the PSL represent boundaries between registrar-type organizations
and their clients. The "owner" of the suffix is one organization, and the "owner"
of a domain label immediately preceeding the suffix is a different organization.

In the case of suffixes in the ICANN section, the "owner" of a suffix is ICANN /
an affiliated registrar. In the case of suffixes in the Private section, the
"owner" of a suffix is a private organization offering a service for clients
to take ownership of subdomains underneath its own ICANN-suffixed domain.

> #### 2. Multiple Suffixes per Domain

Reviewer 3: There can be multiple suffixes per domain but the distinction between private section and ICANN section is a bit arbitrary. There could be multiple entities acting as "registrars" in a domain.

Author: Yes you're right. There is one distinction between the two sections: every Private-section suffix *always* has a shorter ICANN-section suffix at the end of it (according to my analysis of the PSL). So Private-section suffixes are always "multiple-suffix"-type domains. I will see if I can update this section to reflect your comment.

Author: Addressed in c2171564bebfbd15314e2e147a613129044a4753: reworded the section to reflect the arbitrary ICANN vs Private distinction, and content-owners vs registrars

> If a domain name ends with a suffix listed in the Private section of the PSL,

Reviewer 3: Maybe stress that in almost all cases you want to use the full PSL. If you're trying to attribute abuse at some point you want to go from "a.foo.com" and "b.foo.com" and "c.foo.com" are all distributing malware to "foo.com" is distributing malware. That's the only reason I can think of, though and maybe that should instead be a generic "attribute to all higher labels".

Author: Yes, the API in this proposal defaults to using the full PSL (i.e. including Private-section suffixes). Use Case 1 & 2 in this proposal note that there may be cases where it can be useful to exclude Private-section suffixes, therefore the option parameter `excludePrivateSuffixes` is provided. I will see if I can incorporate your suggested use-case in the document.

Author: Addressed in c2171564bebfbd15314e2e147a613129044a4753: changed to always use the full PSL, noted the use case in the "Future Work" section

then it must also end with a shorter suffix in the ICANN section. The question
of which one should be taken to be the domain's public suffix depends on the
specific use case.

##### Example

> | Domain name         | Private suffix  | ICANN suffix |

Reviewer 3: It's called a public suffix so calling it a private suffix here might be a bit confusing. `Private section suffix`?

Author: Sure, I will update this

Author: Addressed in c2171564bebfbd15314e2e147a613129044a4753

|:-------------------:|:---------------:|:------------:|
| foo.bar.wixsite.com | wixsite.com     | com          |

#### 3. Known vs Unknown

If a domain name does not match any suffix in the PSL, it is considered to have
an unknown suffix. Depending on the use case, the public suffix could be taken
by default to be the last domain label of the domain name, or alternatively
the domain name could be considered invalid.

> **Note:** it may be more performant to allow unknown suffixes and assume a single-label

Reviewer: I'm not too concerned about the performance aspect of this; any IPC to handle the API call is likely more expensive than a look up in the PSL.

The proposal does not describe why unknown suffixes should be supported. Could you elaborate when one would want and when one does not want public suffixes? And offer a recommendation on default behavior that is not dependent on (likely not relevant) performance considerations?

It's probably worth pointing out explicitly that there may be web pages displayed in the browser that are not on a registrable domain, e.g. when a local intranet has custom non-public host names.

Author: I will update this section regarding unknown suffixes as you suggest.

On the performance point: I am concerned about performance of this API, because if it turns out to be slower than extensions' own public suffix implementations, then they may not use this API. In particular, Use Cases 1 (Filter Requests by Organization) and 3 (Detect Third-Party Requests) in this proposal are performance-sensitive, because they obtain the registrable domain for every request (not just top-level pages) as the user browses.

Reviewer 3: If it's faster for the extension to use an internal list, wouldn't that imply that we should really improve our implementation?

Author: Extensions that use their own internal list benefit from not having to do a cross process call to get the public suffix, because everything stays within the extension itself. This could theoretically make the extensions' own hand-rolled implementations faster. The downside is, they are duplicating work already done in the parent browser.

Reviewer 3: My gut feeling says this doesn't matter but I have no data on it.

Author: Addressed in c2171564bebfbd15314e2e147a613129044a4753: mentioned intranet hostnames and changed treatment of unknown eTLDs

suffix by default, because it allows the following optimisation of the lookup algorithm:
all single-label suffixes in the PSL can be excluded from the lookup, since they do not
need to be matched specifically.

#### 4. IDN

The PSL uses Unicode, and contains International Domain Name (IDN) suffixes. Depending
on the use case, when returning the *registrable domain / eTLD+1* for a domain name,
either Unicode or Punycode may be preferred.

> **Note:** the [PSL algorithm](https://github.com/publicsuffix/list/wiki/Format#formal-algorithm)

Reviewer 3: I think this might be a made up requirement. (That wiki page really needs some work.) The algorithm is pretty trivial and just needs to compare labels for equality so any encoding should be fine.

Author: I will note this in the document.

Author: Addressed in c2171564bebfbd15314e2e147a613129044a4753

* allows the user to choose which domain names the extension should act on,
and which should be ignored by the extension

In effect, such extensions allow users to create request filtering rules that
restrict the extension to acting on only web requests having specific domain names.

When choosing domain names for the request filter, users may want to be able to choose
all domain names for an organization without having to enter them each in individually.
The user may want to specify request filtering rules as follows:

| # | Filtering rule type | User value | Examples of domains affected |
|:-:|---------------------|:----------:|-----------------------------:|
| 1 | organizational base domain and all possible subdomains | myorg.co.uk | myorg.co.uk<br>subdomain1.myorg.co.uk<br>subdomain2.myorg.co.uk |
| 2 | organization name and all possible suffixes | myorg | myorg.co.uk<br>myorg.com<br>myorg.net |

To support this use case, such extensions may:

* automatically calculate the set of organization names and *registrable domain / eTLD+1*s
from a user-specified set of domain names, and propose these as filtering rules for
the user to choose

> * prevent users from mistakenly specifying a *public suffix / eTLD* as a filtering rule,

Reviewer 3: This would probably be a mistake. There are unresolved, lengthy, ancient discussions on this elsewhere but basically, at the moment, public suffixes sometimes have websites.

Author: Noted - I will reword this.

Author: Addressed in c2171564bebfbd15314e2e147a613129044a4753

believing it to be a *registrable domain / eTLD+1*, e.g. `co.uk` instead of `myorg.co.uk`
* apply the users' filtering rules automatically while the user browses by determining
the *registrable domain / eTLD+1* of each web request, and testing this using appropriate
regexes to decide whether or not to act on the request

##### PSL Requirements

| PSL Feature            | Requirement | Discussion |
|------------------------|-------------|------------|

> | Allow Private Suffixes | Yes & No    | some filters may require private suffixes, others ICANN-only |

Reviewer 3: The PSL exists because filtering by organization requires the information on the PSL so only using the ICANN section seems just wrong here.

Reviewer 3: Another reason why I dislike this is, that the distinction between an ICANN/IANA TLD and an eTLD is - in practice - poorly defined.
The "official" ICANN/IANA TLD is only the last label. There is an [IANA list](https://www.iana.org/domains/root/db) and another [list for gTLDs](https://www.icann.org/resources/registries/gtlds/v2/gtlds.json).

However, the PSL is trying to gather information from the official registrars and include that. To maintain the PSL we try to get in touch with the registrars to get them to update their "official" subdomains or to ask questions but we often fail.

Look at  the`.au` section of the list, they cannot agree what their TLDs are but they are in the ICANN section. `.bd` even has a wildcard `*.bd` entry in the ICANN section. `.co.uk` is in the ICANN section but the actual IANA TLD is still only `.uk`.

Author: I will update the proposal to note the issues you point out regarding the PSL sections.

Author: Addressed in c2171564bebfbd15314e2e147a613129044a4753

|                        |             | preventing user mistake may require both types of suffix |

> | Allow Unknown Suffixes | Yes         | provides better performance |

Reviewer 3: Afaiu this would only make a difference when given a single label. But a single label is either not a valid URL or it is a TLD. If you're trying to look up `isValidURL(...)` you don't want the optimization because it breaks the use-case. If you're looking up `isSameSite(...)` then the optimization for a single label is irrelevant because you have to compare them anyway.

Author: The PSL algorithm starts with a candidate domain name and removes labels in turn until a matching suffix is found. For any unknown-suffixed candiate domain (regardless of how many labels it has), the algorithm will always reach the final label without finding any matches in the PSL. At this point, if we are allowed to assume that all unknown single-label suffixes are valid, then the algorithm can:

1. avoid doing the final check against the PSL using the final label. E.g. `foo.bar.baz` (check) --> `bar.baz` (check) --> `baz` (no check needed). This may save a few CPU cycles for every candidate domain lookup.
2. avoid storing *any* single-label suffixes. This allows a possible reduction in browser startup time since we are loading fewer PSL suffixes into memory from disk, and thereafter lower browser memory usage due to holding fewer PSL suffixes in memory.

Author: Addressed in c2171564bebfbd15314e2e147a613129044a4753: noted the possible optimization but that it may be not significant

|:----:|----------------------------:|----------------------------:|:-----------:|:------------:|
| #1   | foo.com                     | foo.com                     | ICANN       |              |
| #2   | bar.com                     | bar.com                     | ICANN       | Yes          |
|      |                             |                             |             |              |
| #1   | foo.amazonaws.com           | amazonaws.com               | ICANN       |              |
| #2   | bar.amazonaws.com           | amazonaws.com               | ICANN       | No           |
|      |                             |                             |             |              |
| #1   | foo.amazonaws.com           | amazonaws.com               | ICANN       |              |
| #2   | bar.us-east-1.amazonaws.com | bar.us-east-1.amazonaws.com | Private     | Yes          |

##### PSL Requirements

| PSL Feature            | Requirement | Discussion |
|------------------------|-------------|------------|
| Allow Private Suffixes | Yes         | including all suffixes in PSL means more information about third-party boundaries |

> | Allow Unknown Suffixes | Yes         | provides better performance |

Reviewer: Why are Unknown suffixes required in this case? Theoretically better performance does not automatically translate to a functional requirement.

Author: Assuming there is no performance benefit from allowing unknown suffixes, I will revisit this requirement.

Author: Addressed in c2171564bebfbd15314e2e147a613129044a4753: changed treatment of unknown eTLDs

from this API due to the reasons stated in the [Objective](#objective) section.

| Extension | Relevant use cases | Description |
|-----------|--------------------|-------------|
| [multi-account-containers](https://addons.mozilla.org/en-US/firefox/addon/multi-account-containers/) | #1, #2 | isolate requests into containers by organizational base domain, view all domains grouped by organization |
| [uBlock Origin](https://addons.mozilla.org/en-US/firefox/addon/ublock-origin/) | #1, #2, #3 | create organization/3rd-party request-blocking rules, view request blocking status by organization, create blocking rules for specifically 3rd-party requests |
| [adblock-plus](https://addons.mozilla.org/en-US/firefox/addon/adblock-plus/) | #1, #3 | create organization/3rd-party request-blocking rules |
| [duckduckgo-for-firefox](https://addons.mozilla.org/en-US/firefox/addon/duckduckgo-for-firefox/) | #3 | exclude requests from tracking detection unless they are 3rd-party |
| [violentmonkey](https://addons.mozilla.org/en-US/firefox/addon/violentmonkey/) | #1 | run scripts automatically for requests having a specific organization name plus any suffix |

## Specification

### Schema

> A new API `publicSuffix` is added as follows:

Reviewer 3: I've dug into PSL use-cases a couple of time now. I think it might be better to make the API
`isValidURL(...)` and `isSameSite(...)`. That should be enough for use-cases (1.) and (3.) above. For use case (2.) you're restricted to comparison based sorting algorithms but that's probably okay in practice.

Reviewer 3: @remusao, if you have feedback on this I would love to hear it!

Reviewer 3: As @Rob--W just pointed out in chat, there are some user-interface use-cases which would still need the `getRegistrableDomain` function. For example, highlighting in the URL bar or giving a name to a group of same-site URLs.

I still think it would be good to have an explicit API for the two cases mentioned above.

Author: We can certainly add `isSameSite(domain1, domain2)` to the proposal. Another name for the function could be `hasSameRegistrableDomain(domain1, domain2)`. It would be equivalent to:

```
await publicSuffix.getRegistrableDomain(domain1) === await publicSuffix.getRegistrableDomain(domain2)
```

Regarding `isValidURL(string)`, another name for this function could be `hasKnownPublicSuffix(domain)`. As you note in another comment, this may not always be reliable due to delays in updating the PSL:

> most browsers currently do not update their PSL, the "made up TLD" situation can also arise when a new TLD gets introduced.

However, as @Rob--W mentions in another comment, Firefox relies on the `hasKnownPublicSuffix(domain)` check when deciding how to handle user input in the navigation bar (i.e. search vs navigate). So the possible unreliability of this function may be acceptable

Reviewer 3: I think your names are better. Mine are based on the use-case and that's probably good for thinking about it but they are wrong in terms of what they actually do.

We should probably align on what we want to call things across the document. I like "eTLD+1" and "eTLD" because they are short and consistent but I agree that "registrable domain" might be clearer. So my vote would be
**eTLD** = Public Suffix
**Registrable Domain** = eTLD+1
And if we find other names for it we should probably state the mapping clearly. I know Firefox source also uses BaseDomain and RootDomain for eTLD+1.

Author: Makes sense, I'll update the proposal to use these terms

Author: Addressed in c2171564bebfbd15314e2e147a613129044a4753: added `hasSameRegistrableDomain()`, changed `getRegistrableDomain()` to return `null` for unknown eTLDs thereby providing an equivalent way of determining `hasKnownPublicSuffix()`, adopted *eTLD* and *Registrable Domain* as standard terms.

```ts
namespace publicSuffix {
  //
  // Object containing both registrable domains (i.e. ICANN/unknown, Private)
  // for a domain.
  //
  interface RegistrableDomains {
    // The original domain whose registrable domain we want
    domain: string,
    // The ICANN-suffixed or unknown-suffixed registrable domain.
    // Null if an error occurred.
    base?: string,
    // The Private-suffixed registrable domain.
    // Null if an error occurred, or if the domain has no matching Private suffix.

>     private?: string,

Reviewer: Do we really need all three options (domain, base, private)?

- (input) `domain` is redundant, because the extension can keep track of the parameters that it had sent to the API.
- Requiring `base` and `private` may require two lookups in the PSL, even if the extension does not need it. An extension interested in both can call the API twice with `excludePrivateSuffixes` set to true and false.

Given this, what do you think of reducing the number of options to just one `registrableDomain`?

Author: Calling the API twice with `excludePrivateSuffixes` set to true and false has the following potential issues:

**Duplication of work**
* Must parse/canonicalize the same input domain parameter on each API call.
* The PSL lookup involves removing each label in turn from the input domain and testing the remaining suffix until a match is found. Therefore on the second API call, the same unmatching candidate suffixes that have already been tested in the first API call will again be retested. The difference will be that the algorithm keeps going further in terms of removing labels with the `excludePrivateSuffixes=true` call.

**Duplication of returned arrays**
* If the registrable domains array obtained with `excludePrivateSuffixes=false` happens to only contain ICANN domains (because no matching private suffixes exist), then the second API call with `excludePrivateSuffixes=true` will return exactly the same registrable domains array again.

Note that my proposal does not require "two lookups". One possible implementation would simply continue removing labels after finding a private suffix until the ICANN suffix is found, i.e. not a full lookup, but a continuation of the current lookup. Even better, a more optimal implementation would have the corresponding ICANN suffix pre-calculated and stored with every private suffix, such that no further work would be needed to determine the ICANN suffix upon matching a private suffix.

I am happy to remove `domain` in the result, because I agree this can be inferred from the array position.

Author: @simon-friedberger @Rob--W I am in the process of updating the proposal to incorporate Simon's feedback about the somewhat arbitrary distinction between the Private and ICANN sections of the PSL. Would we prefer to avoid using the term "private" in this proposal's API given those issues?

In its current form, this API method `getRegistrableDomains()` returns an object containing registrable domain values as follows:

| Field --> | `base` | `private?` |
|-:|-:|-:|
|  | ICANN section / unknown | Private section |
| foo.bar.wixsite.com | wixsite.com | bar.wixsite.com |
| xyz.example.com | example.com |  |

If the Private-section vs ICANN-section distinction is arbitrary, should we reconsider these fields? An alternative terminology could be:

| Field --> | `default` | `short?` |
|-:|-:|-:|
|  | Private section / ICANN section / unknown | ICANN section |
| foo.bar.wixsite.com | bar.wixsite.com | wixsite.com |
| xyz.example.com | example.com |  |

The `short` registrable domain would only be populated if the `default` registrable domain had an eTLD from the Private section (all Private section eTLDs are longer).

Similarly, we could rename the `excludePrivateSuffixes` key of the options parameter for the non-batching API method `getRegistrableDomain()` to simply `short`.

| Option --> | `short=false` | `short=true` |
|-:|-:|-:|
|  | Private section / ICANN section / unknown | ICANN section |
| foo.bar.wixsite.com | bar.wixsite.com | wixsite.com |
| xyz.example.com | example.com |  |

Reviewer 3: I still think we should not offer the "short" version at all. I don't see a good argument in the use-cases for having this. Did I miss something?

Author: Addressed in c2171564bebfbd15314e2e147a613129044a4753: API no longer provides a way of getting the "short" registrable domain or distinguishes between ICANN vs Private-section eTLDs

    punycode: boolean,
  }

  //
  // Gets the longest registrable domain for a specified domain.
  //
  // Example:
  //
  // let domain = await browser.publicSuffix.getRegistrableDomain("www.example.co.uk");
  // ==> "example.co.uk"
  //
  export function getRegistrableDomain(
    // The domain name whose registrable domain we want to find
    domain: string,
    // Options that control the behaviour of the lookup algorithm
    options?: RegistrableDomainOptions,
  )
  // Resolves to the longest registrable domain of the input domain name
  : Promise<string>;

  //
  // Gets both registrable domains (i.e. ICANN/unknown, Private) for each domain name
  // in an array of domain names.
  //
  // Example:
  //
  // let domains = await browser.publicSuffix.getRegistrableDomains([
  //   "foo.bar.wixsite.com",
  //   "www.example.net",
  //   "a..b",
  // ]);
  // ==> [
  //   { domain: "foo.bar.wixsite.com", base: "wixsite.com", private: "bar.wixsite.com" },
  //   { domain: "www.example.net",     base: "example.net",                            },
  //   { domain: "a..b",                error: "Invalid domain name",                   },
  // ]
  //

>   export function getRegistrableDomains(

Reviewer 3: Are callers gaining anything from calling this with an array instead of calling the other function multiple times?

Author: The problem with calling `getRegistrableDomain` multiple times is there is overhead associated with an extension calling an async function on the parent browser. If you have a list of 50 domains, you would be making 50 async calls to the parent browser. However, with the `getRegistrableDomains` batching approach, you would be making a single async call to the parent browser passing all 50 domains at once.

I hacked a quick mockup of the two approaches using a simplified implementation of the Public Suffix API in a modified Firefox, and the batched approach seems to be about 2-3 times faster for 50 domains.

In addition, `getRegistrableDomains` returns *both* the ICANN section suffix and the Private section suffix for each input domain, whereas `getRegistrableDomain` returns only one or the other (according to `excludePrivateSuffixes` in the `options` input parameter).

This function `getRegistrableDomains` is for Use Case "2. Group Domains in UI" in this proposal.

Author: Addressed in c2171564bebfbd15314e2e147a613129044a4753: noted the performance justification for batching

    // The domain names whose registrable domains we want to find
    domains: Array<string>,
    // Options that control the behaviour of the lookup algorithm
    options?: RegistrableDomainOptions,
  )
  // Resolves to the registrable domains of each input domain name
  : Promise<Array<RegistrableDomains>>;

  //
  // Gets the PSL dataset version if available
  //

>   export function getVersion(): string?;

Reviewer 2: @simon-friedberger let me know that the PSL has been updated with a `VERSION` metadata item like this: https://github.com/publicsuffix/list/issues/1808#issuecomment-2455793503

Should we mention that here and agree on using that as the version we return?

Reviewer: I'd be fine with that.

I am also fine with leaving the version method off altogether if it were to be a blocker to approving the proposal.

After all, there are many existing APIs that are dependent on the result of the public suffix list without exposing versioning information.

Reviewer 2: I definitely agree that this shouldn't be a blocker. However, I hope it will be non-controversial. I expect most (likely all) browser vendors don't have easy access to this information today but it should be fairly trivial to get access to.

Author: Addressed in c2171564bebfbd15314e2e147a613129044a4753: included the VERSION metadat reference

However, if an `options` object is passed to `getRegistrableDomain()` with key
`excludePrivateSuffixes` set to `true`, then Private (non-ICANN) suffixes should be
**excluded** from the lookup algorithm.

##### Example

`domain` = foo.bar.wixsite.com

| Option                                   |   Registrable Domain | PSL Section |
|------------------------------------------|:--------------------:|:-----------:|
| excludePrivateSuffixes = false (default) |      bar.wixsite.com | Private     |
| excludePrivateSuffixes = true            |          wixsite.com | ICANN       |

#### 2. Multiple Suffixes per Domain

The lookup performed by `getRegistrableDomain()` should select the **longest** matching
suffix (unless specifically excluded using the `excludePrivateSuffixes` option).

##### Example

`domain` = foo.bar.lib.de.us

| Candidate Suffix | PSL Section |
|-----------------:|:-----------:|
|            de.us | ICANN       |
|        lib.de.us | Private     |

The longest is lib.de.us, so `getRegistrableDomain()` resolves to bar.lib.de.us

#### 3. PSL Special Rules

The lookup performed by `getRegistrableDomain()` should adhere to the

> [PSL algorithm](https://github.com/publicsuffix/list/wiki/Format#formal-algorithm).

Reviewer 3: I support doing it this way but we will have to talk about it because that is not what browsers are currently doing. 

That situation has been unresolved for a long time and people have been working around it or ignoring it. I suggest we simply defer to what the browsers are already doing here because that is what we want.

Author: I will incorporate your comment in the proposal.

Author: Addressed in c2171564bebfbd15314e2e147a613129044a4753: noted the situation re. browsers

|-------------------------:|--------------:|-----------------:|:--------------:|-------------------:|
| sub.domain.com           | com           | com              | Simple rule    | domain.com         |
| sub.domain.co.uk         | co.uk         | co.uk            | Simple rule    | domain.co.uk       |
| sub.domain.gov.ck        | gov.ck        | *.ck             | Wildcard rule  | domain.gov.ck      |
| sub.domain.any.ck        | any.ck        | *.ck             | Wildcard rule  | domain.any.ck      |
| sub.sub.domain.any.ck    | any.ck        | *.ck             | Wildcard rule  | domain.any.ck      |
| www.ck                   | ck            | !www.ck          | Exception rule | www.ck             |
| sub.www.ck               | ck            | !www.ck          | Exception rule | www.ck             |
| sub.sub.www.ck           | ck            | !www.ck          | Exception rule | www.ck             |

#### 4. Unknown Suffixes

If no matching suffix is found in the PSL for a `domain` parameter, then unless it is determined
to be specifically [invalid](#6-invalid-domain-parameter), it should be assumed the domain has a

> single-label suffix.

Reviewer: What is the source for the required assumption of "single-label" suffix?

This proposal does not include an example where known vs unknown matters, but in Firefox specifically, there is at least one example where it matters (https://bugzilla.mozilla.org/show_bug.cgi?id=1621168): In determining whether to issue a search query or whether to try a navigation, a PSL lookup is made:
- If valid, attempt to navigate.
- If invalid, use search engine.

Unknown entries in the PSL should also trigger a search query, but unconditionally making it return a single label would rule out that use.

Author: I am happy to remove this assumption, and instead add an option to the API allowing users to opt-in to this behaviour of assuming a single-label suffix. I was guided in my analysis by the following:

- Google Chrome's Public Suffix implementation makes this assumption, I believe. See [GetDomainAndRegistry](https://source.chromium.org/chromium/chromium/src/+/main:net/base/registry_controlled_domains/registry_controlled_domain.h;l=184;drc=b3929870e1504f3937609eff00b1745fbd38916d): "If no matching rule is found in the effective-TLD data (or in the default data, if the resource failed to load), the last subcomponent of the host is assumed to be the registry."
- [tltds](https://github.com/remusao/tldts) also makes this assumption in its implementation. (tltds is a javascript library for obtaining Public Suffixes that is used by popular extensions such as [bitwarden password manager](https://addons.mozilla.org/en-US/firefox/addon/bitwarden-password-manager/) and [violentmonkey](https://addons.mozilla.org/en-US/firefox/addon/violentmonkey/).) However, there is also an open [tltds issue](https://github.com/remusao/tldts/issues/2121) caused by this assumption.
- Use Cases 1 (Filter Requests by Organization) and 3 (Detect Third-Party Requests) in this proposal are performance-sensitive, because they obtain the registrable domain for every request (not just top-level pages) as the user browses. I wanted to reduce the chance of making users' browsers feel permanently slower after installing these sorts of extensions. This assumption of a single-label suffix by default may allow a faster implementation due to avoiding an explicit PSL lookup for all single-label TLDs.
- Use Case 2 (Group Domains in UI) in this proposal may benefit from this assumption, because without it all unknown-suffixed domains would be grouped together.

Reviewer 4: Hey, I'm the maintainer of `tldts` and wanted to clarify that the [issue linked](https://github.com/remusao/tldts/issues/2121) is about `tldts-experimental`, a faster but less accurate implementation of `tldts`. The main `tldts` package (which most projects are using) is fully accurate, and behaves in a slightly different way by returning a single label in case no rule was found matching the last label (or anything more specific), but will also return two flags: `isIcann` and `isPrivate`, which allow to know where the matching rule came from (i.e. the ICANN  or PRIVATE section of the public suffix list). If both of these flags are `false`, then we can assume that the fallback behavior of using the last label as public suffix was triggered. For example:
```sh
$ npx tldts https://example.thissuffixdoesnotexist
{
  "domain": "example.thissuffixdoesnotexist",
  "domainWithoutSuffix": "example",
  "hostname": "example.thissuffixdoesnotexist",
  "isIcann": false,
  "isIp": false,
  "isPrivate": false,
  "publicSuffix": "thissuffixdoesnotexist",
  "subdomain": ""
}
```

I lack context on this spec but this might allow to combine the single-label suffix fallback behavior with the use-cases outlined by @Rob--W above.

Reviewer 3: The main use-cases for the PSL are `isValidUrl(...)` which does not want this behavior (because you want to know if the TLD is known) and `isSameSite(...)` for which it *almost* doesn't matter.

It only helps if you have made up TLDs. `printer.homenet` and `backup.homenet` would both have an empty eTLD and eTLD+1 would be "homenet". If you assume the last label must be a TLD then you end up with the correct behavior.

However, since most browsers currently do not update their PSL, the "made up TLD" situation can also arise when a new TLD gets introduced.

I think the correct solution here is to make the API `isValidURL(...)` check the last label and `isSameSite(...)` to assume that it is a valid TLD. But maybe there are use-cases that I'm not seeing which need a third option.

Author: Addressed in c2171564bebfbd15314e2e147a613129044a4753: changed treatment of unknown eTLDs, added `hasSameRegistrableDomain()`, mentioned Firefox search vs navigate use case

##### Example

| Domain parameter       | Registrable domain | PSL Section |
|-----------------------:|-------------------:|:-----------:|
| www.example.foobar     | example.foobar     | n/a         |
| www.example.co.foobar  | co.foobar          | n/a         |

#### 5. IDN

The `domain` parameter passed to `getRegistrableDomain()` may be either Unicode
or Punycode.

When settling the promise returned by `getRegistrableDomain()`, the resulting

> domain name should be converted to Unicode from Punycode by default.

Reviewer: Why should the API convert the domain to Unicode by default? When the input is the host name from a URL, it would be reasonable to expect a valid hostname. I'd argue that punycode is the more sane default.

Extensions can bundle the custom library/logic to convert punycode to unicode if desired, because the algorithm is well known and fixed.

Author: I can change this to returning punycode by default, with an API option to convert to unicode. Unicode registrable domains are required by Use Case 2 (Group Domains in UI) in this proposal.

Author: Addressed in c2171564bebfbd15314e2e147a613129044a4753

However, if an `options` object is passed to `getRegistrableDomain()` with key
`punycode` set to `true`, then Punycode should be used instead.

##### Example

`domain` = foo.bar.example.مليسيا

| Option                     |     Registrable Domain |
|----------------------------|-----------------------:|
| punycode = false (default) |         example.مليسيا |
| punycode = true            | example.xn--mgbx4cd0ab |

> #### 6. Invalid domain parameter

Reviewer 3: This is important to keep in mind.`isSameSite("100.200.30.2", "100.200.31.2")` should probably return false so even if the last label is assumed to be a TLD (and thus not checked) there should be an exception that checks for IPv4 addresses.

Author: Noted, I will include this point.

Author: Addressed in c2171564bebfbd15314e2e147a613129044a4753: included the example showing why IP addresses should be rejected

The promise returned by `getRegistrableDomain()` should reject if the `domain` parameter
meets any of the following criteria:
* Contains a character that is invalid in an Internationalized Domain Name (IDN) - e.g. symbols, whitespace
* Is an IP address - IPv4 or IPv6

> * Is a public suffix itself - including the case of it being a single-label suffix not explicitly matched in the PSL

Reviewer 3: I think it would be good to think about the callers a bit more to determine what they want here.
If we assume the caller is something like `isKnownPublicSuffix(...)` then they would want a "true" for IPs and things that are already a public suffix but a false for empty labels or the root label. If we assume the caller is something like `getHighlightParts(...)` then an IP should probably look like a registrable domain.

Author: I have updated the API in c2171564bebfbd15314e2e147a613129044a4753: `getRegistrableDomain()` is now a proxy for `hasKnownPublicSuffix()`, because it returns `null` in cases where a domain is valid but has an unknown eTLD (i.e. it no longer defaults to assuming a single-label eTLD). So if the promise returned by `getRegistrableDomain()` fufills with a nonnull value, then it can be inferred that the domain has an eTLD in the PSL.

The API does not currently have a `getHighlightParts()` method, but we could possibly add it in the Future Work section if we can identify an appropriate use case.

Reviewer 3: I'm not sure if I was sufficiently clear here.

My point is there should be an API such that if I want to highlight the different URL parts in the address bar or find out what the current "site" is for applying some setting there needs to be an API that takes.

`docs.google.com/...` -> `google.com` - ICANN 
`foo.github.io` -> `foo.github.io` - PRIVATE
`10.0.0.1/login` -> `10.0.0.1` - even though this is an IP
`avocado.banana` -> `avocado.banana` - even though this is not a valid TLD, because it can be made to resolve locally

On the other hand, if the extension is trying to determine if something is a domain or a search term we probably want
`docs.google.com/...` -> `google.com` - ICANN 
`foo.github.io` -> `foo.github.io` - PRIVATE
`10.0.0.1/login` -> `10.0.0.1` - even though this is an IP
`avocado.banana` -> ERROR/searchTerm

I think having different functions to call depending on what people want will probably be a nicer API but just returning a tuple and adding additional information about the result like `type:RegistrableDomain|IPAddress|UnknownTLD` would also be an option.

Author: Yes I can see how that could be useful. Is it something we could address in a future version of this API?

Reviewer 3: It doesn't seem like a lot of effort to me so I would say let's make sure to include it in the first version. (Otherwise people will always use the old version because "what if the new version isn't supported yet?")

Author: Your two proposed use cases could be implemented using the existing API in this proposal as follows:

```javascript
async function getIPAddressOrRegistrableDomain(domain) {
  // IP Address (note: this regex is not perfect!)
  if (/\d\d?\d?\.\d\d?\d?\.\d\d?\d?\.\d\d?\d?/.test(domain)) {
    return domain;
  }
  // Registrable Domain (returns null if unknown, throws error if invalid)
  return publicSuffix.getRegistrableDomain(domain);
}

try {

  // Use case 1: what is the curent site?
  const currentSite = await getIPAddressOrRegistrableDomain(domain) || domain;

  // Use case 2: is this a known domain, or should I search?
  const isDomain = await getIPAddressOrRegistrableDomain(domain) !== null;

} catch (e) {
  // Invalid domain
}
```

Reviewer 3: Sorry, for not responding! I'm not 100% sure what you mean here but this does not cover IPV6 addresses. And how should this treat domains which are not registrable but intended to be suffixes like home.arpa? And what does getRegistrableDomain return on things which are not registrable but by the "single label rule" count as a TLD, like feeding it "localhost"?

Should we maybe just start prototyping this somewhere and write a bunch of tests for edge cases and see, for example, how it maps to the use in common extensions and the internal use of https://searchfox.org/mozilla-central/source/netwerk/dns/nsIEffectiveTLDService.idl in Firefox?

(I think you already did that!)

Author: The reason I showed this example code is to demonstrate:

1. Extensions have everything they need using the existing API in this proposal to cover your use cases.
    
    They still need to write their own code to pre-parse for IP addresses, but this is a more general issue not specific to this API, since browsers do not currently offer anything built-in to do this for developers. For example, here is a webpage explaining how to [parse IP addresses with javascript](https://www.geeksforgeeks.org/extracting-ip-address-from-a-given-string-using-regular-expressions/).

2. Adding functionality to this API to determine if a string is an IP address (whichever type) may be making it more general-purpose than it should be.

    Perhaps it should be left up to the developer to filter out IP addresses before calling this API. Alternatively, functionality to parse IP addresses could be offered in a different API. It would certainly be useful, as [this stackoverflow question](https://stackoverflow.com/questions/4460586/javascript-regular-expression-to-check-for-ip-addresses) and [this stackoverflow question](https://stackoverflow.com/questions/5284147/validating-ipv4-addresses-with-regexp) show.

In answer to your questions:

> this does not cover IPV6 addresses

It was not meant to, the regex I gave was to point out that it would be up to extension developers to come up with a solution for pre-parsing IP addresses before calling this API.

> how should this treat domains which are not registrable but intended to be suffixes like home.arpa

The API returns `null` if a string is a domain with an unknown suffix. This is used in my code example here:

```
  // Use case 1: what is the curent site?
  const currentSite = await getIPAddressOrRegistrableDomain(domain) || domain;
```

In this code snippet, in the event the API returns `null`, the code substitutes the original domain string instead - i.e. it assumes it is an unknown local domain. (This is simplified code: the extension developer may want to do extra work to get the last label from the `domain` string instead of just using the string as-is.)

> And what does getRegistrableDomain return on things which are not registrable but by the "single label rule" count as a TLD, like feeding it "localhost"?

An earlier version of this proposal stated that the "single-label rule" should apply in the case of domains with unknown suffixes. However, following reviewer feedback, this assumption was removed. Instead, where a domain has an unknown suffix, the API returns `null`. It is then left to the extension developer to decide what to do, e.g. to assume a single-label suffix.

> Should we maybe just start prototyping this somewhere and write a bunch of tests for edge cases and see, for example, how it maps to the use in common extensions and the internal use of https://searchfox.org/mozilla-central/source/netwerk/dns/nsIEffectiveTLDService.idl in Firefox?

Yes, use cases identified in common extensions are set out in this proposal. Perhaps you could review the table in section "7. Summary of behaviours" in this proposal, since it may help demonstrate how the API handles the various edge cases?

Reviewer 3: I agree with you, that the suggested API can be used to build what is necessary.

But on the other hand, this is a browser API it will very often get a "location" and have to figure out what the "base domain" is. So why would we make each developer sort out the edge cases themselves again? I would prefer an API which is more aligned with what developers need.

Author: Maybe you could suggest the API method you have in mind? (I.e. method name, parameter(s), return type)

Author: In the current proposal, an error is thrown if the value passed to `getRegistrableDomain(value)` is invalid, e.g. if it contains invalid characters or is an IP address.

One way of supporting your use case would be to make it possible to determine if the error was thrown due to the value being an IP address. We could add a requirement that the [`cause`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/cause) property of the error should be set to `{ code: "IPAddress" }`.

Then, my earlier example code to handle your use cases would be updated as follows:

```javascript
async function getIPAddressOrRegistrableDomain(domain) {
  try {
    return await publicSuffix.getRegistrableDomain(domain);
  } catch (e) {
    if (e.cause && e.cause.code === "IPAddress") {
      return domain; // It's an IP address
    }
    throw e; // Not an IP address, so rethrow the error
  }
}

try {

  // Use case 1: what is the curent site?
  const currentSite = await getIPAddressOrRegistrableDomain(domain) || domain;

  // Use case 2: is this a known domain, or should I search?
  const isDomain = await getIPAddressOrRegistrableDomain(domain) !== null;

} catch (e) {
  // Invalid domain
}
```

Reviewer 3: Seems good. I think this function should be part of the API. And calling this function `export function hasSameRegistrableDomain(...domains: string[]) : Promise<boolean>;` should also allow matching same IPs.

Author: Ok sure, but we may need to think of a different function name than `hasSameRegistrableDomain()`, because an IP address is not quite the same as a registrable domain.

Also, the name of this API is `publicSuffix`, but we are now adding methods that explicitly support IP addresses, which do not really have anything to do with public suffixes. Are we happy to keep the name `publicSuffix` for the API?

I'm not saying we can't make the changes you're suggesting, I just want to make sure we're ok with the naming of the API and methods.

Reviewer 3: We could make a higher level function like "getSite" and "hasSameSite" which just allows the things that make sense in a browser URL bar, like an IP or we just add an argument "allow IP". I guess the first one is nicer because it allows updating that function if we think of something else that needs to be allowed.

Reviewer: To make sure we stay on track.

The objective is offering an API that exposes the relevant hard-to-replicate behavior of the public suffix list.

IP addresses do not have to be recognized and labeled specially in the API if it doesn't fit naturally otherwise.

If the input to getRegistrableDomain is an IP address, I would expect the result to be null. That is also consistent with the "registrable domain" description at https://url.spec.whatwg.org/#host-miscellaneous

By returning null, the API can easily be used as follows:

```javascript
// E.g. with tabs.Tab type from extension API
let host = new URL(tab.url).hostname;
let domain = await browser.publicSuffix.getXXX(host);
let shortestDomain = domain || host;
// Now do something with shortestDomain
```

If someone really wanted to special case IP addresses separately, they can have static logic in the extension itself, without a new extension API.

Author: @robwu Agreed. I will update the proposal so that `getRegistrableDomain(hostname)` returns `null` if the parameter is an IP address, instead of throwing an error.

@simon-friedberger We could add `"OrganizationHostname"` methods to the API as follows:

| Semantic Unit | Can Have Unknown eTLD? | Can Be IP Address? |
| :-: | :-: | :-: |
| RegistrableDomain | No | No |
| OrganizationHostname | Yes | Yes |

#### Examples

| API Method | Param1 | Param2 | Returns |
| -: | :-: | :-: | :-: |
| hasSameRegistrableDomain | avocado.banana | avocado.banana | false |
| isSameOrganization | avocado.banana | avocado.banana | true |
| | | |
| hasSameRegistrableDomain | 1.2.3.4 | 1.2.3.4 | false |
| isSameOrganization | 1.2.3.4 | 1.2.3.4 | true |
| | | |
| getRegistrableDomain | avocado.banana | | null |
| getOrganizationHostname | avocado.banana | | avocado.banana |
| | | |
| getRegistrableDomain | 1.2.3.4 | | null |
| getOrganizationHostname | 1.2.3.4 | | 1.2.3.4 |

Reviewer 3: @mckenfra I like that. Having one function which gives the plain result for the included PSL and a function which does the reasonable thing that users probably want. Which makes a lot of sense given that any PSL query is always heuristic in nature.

I imagine the API would be simpler to use if we just made one function and returned additional information. The result could be { "same": true, "registrableDomain": "avocado.banana", "section": null } or { "same": true, "registrableDomain": "1.2.3.4", "section": null } and it would be obvious which values to use but if somebody wants to be strict they can look at the section value and change the behavior for values which are not actually on the list.

Author: @simon-friedberger Ok great, I will update the proposal.

Author: @Robwu @simon-friedberger I'm just finishing up the changes, but would appreciate your input regarding the shape of the API. Below are two possibilities, let me know if you have a preference for either one (I prefer option 2):

```js
// Option 1
publicSuffix.hasSameRegistrableDomain(hostname1, hostname2)
publicSuffix.getRegistrableDomain(hostname)
publicSuffix.getRegistrableDomains([hostname1, hostname2, hostname3])

publicSuffix.hasSameOrganizationalHostname(hostname1, hostname2);
publicSuffix.getOrganizationalHostname(hostname)
publicSuffix.getOrganizationalHostnames([hostname1, hostname2, hostname3])

// Option 2
publicSuffix.registrableDomain.isEqual(hostname1, hostname2)
publicSuffix.registrableDomain.get(hostname)
publicSuffix.registrableDomain.getAll([hostname1, hostname2, hostname3])

publicSuffix.organizationalHostname.isEqual(hostname1, hostname2)
publicSuffix.organizationalHostname.get(hostname)
publicSuffix.organizationalHostname.getAll([hostname1, hostname2, hostname3])
```

As a reminder, an *organizational hostname* differs from a *registrable domain* in that it can be an IP address or have an unknown eTLD.

Reviewer 3: Don't call it "organizationalHostname", though. It's clearly the same thing as the registrableDomain just a little more lenient. Calling it something else is confusing.

Author: I'm hesitant to use the term *registrable domain* to cover IP addresses and hostnames with unknown eTLDs because:

* an IP address is not a domain, but it *is* a hostname. See the [hostname property](https://developer.mozilla.org/en-US/docs/Web/API/URL/hostname) of Javascript's URL object.
* a hostname with an unknown eTLD is most likely on a home/company *intranet*, whereas domains must be on the *internet*
* IP addresses and intranet hostnames (with unknown eTLDs) are not really "registrable"

Reviewer 3: I can't come up with a better name either. Should we add a 'strict' parameter that can be set to false?

Anybody who knows they definitely are navigating somewhere can set 'strict=false' and will get something that might be the domain.

Thinking about use-cases to validate a bit:

### Use-case: Check if a string looks like a TLD
Set 'strict=true', get none if it doesn't

### Use-case: Highlight eTLD+1, using `get..`
Set 'strict=false', get an approximation of the eTLD+1

### Use-case: Site-specific data (e.g. password autofill, cookies) using `hasSame...` or using the result of `get...` as a key
Set 'strict=false', the result
* matches correctly for IPs
* matches correctly for unknown TLDs
* maybe matches incorrectly for actual hostnames like `www.server1` and `oldwww.server1`. We have no way of fixing this. There is even a setting for controlling the qualification of the DNS lookup called `ndots` which is based on how many dots a name already has.

Author: How about this instead:

```js
// eTLD
publicSuffix.get(hostname)
publicSuffix.getAll([hostname1, hostname2, hostname3])
publicSuffix.parse(hostname)
publicSuffix.parseAll([hostname1, hostname2, hostname3])

// eTLD+1
publicSuffix.registrableDomain.get(hostname)
publicSuffix.registrableDomain.getAll([hostname1, hostname2, hostname3])
publicSuffix.registrableDomain.parse(hostname)
publicSuffix.registrableDomain.parseAll([hostname1, hostname2, hostname3])
```

The `get*` methods return:
* string if the hostname has a known eTLD
* null for everything else (including IP addresses)
* error if the hostname is invalid

The `parse*` methods return the following object:

```js
{
  value: string | null,
  type: string
}
```

`value` = string if the hostname has a known eTLD, else null.
`type` = one of the following:

* KnownSuffix
* UnknownSuffix
* NoSuffix (cannot get registrable domain from a single-label hostname)
* IPAddress
* InvalidHostname 

We get rid of the `hasSame*` methods completely, since it is not very clear what the return value would mean for IP addresses and unknown suffixes.

Author: We can slightly tweak the `publicSuffix.registrableDomain.parse*` methods so that the`value` in the resulting object is populated in more cases instead of being `null`, i.e.:

| `type` | `value` |
| - | - |
| KnownSuffix | eTLD+1 |
| UnknownSuffix | eTLD+1 |
| IPAddress | IP address (i.e. the input hostname itself) |
| NoSuffix | null |
| InvalidHostname | null |

Then it is easy enough to implement a method such a `isSameSite()` as follows:

```js
const [a, b] = await publicSuffix.registrableDomain.parseAll([hostname1, hostname2]);
const isSameSite = a.value && b.value && a.value === b.value;
```

This also leaves flexibility for extension developers to choose to exclude IP addresses / unknown suffixes if they prefer when comparing hostnames:

```js
const [a, b] = await publicSuffix.registrableDomain.parseAll([hostname1, hostname2]);
const hasSameRegistrableDomain =
  a.type === "KnownSuffix" && b.type === "KnownSuffix" && a.value === b.value;
```

Reviewer 5: > My point is there should be an API such that if I want to highlight the different URL parts in the address bar or find out what the current "site" is for applying some setting there needs to be an API that takes.

Do you want to know what's on the PSL, or do you want to know what `<current browser>` treats as a "site"? If it's the latter (seems most useful) then use "Site" in the API name. And it's OK if that API returns things that are NOT on the PSL (like some of your examples) because your extension wants to segregate/decide things based on how `<current browser>` defines a site.

It would be better if "site" information were available in a regular DOM API, not limited to extensions. For example, the [URL interface](https://developer.mozilla.org/en-US/docs/Web/API/URL) already has `origin`, it could have `site`, too. If that becomes true then you wouldn't need to duplicate it on a web extension API that returns detailed information about the PSL. "Site" is derived from the PSL, but they aren't themselves part of the PSL.

Reviewer 3: I like Dan's suggestion of having a "Site" API but we should make sure to have enough tests to make it clear where that behavior is browser dependent.

> Then it is easy enough to implement a method such a `isSameSite()` as follows:
> 
> ```js
> const [a, b] = await publicSuffix.registrableDomain.parseAll([hostname1, hostname2]);
> const isSameSite = a.value && b.value && a.value === b.value;
> ```

Maybe we should start writing down some test cases because I don't think this is quite right. Using the same hostname twice should clearly be same-site so you would want to handle NoSuffix like IP addresses.

Author: Ok, we can change the object returned by `parse*` as follows:

| `type` | `value` |
| - | - |
| KnownSuffix | eTLD+1 |
| UnknownSuffix | eTLD+1 |
| IPAddress | the input hostname itself (an IP address) |
| NoSuffix | the input hostname itself (a single label) |
| InvalidHostname | null |

Then this code should work as you want it to:

```js
const [a, b] = await publicSuffix.registrableDomain.parseAll([hostname1, hostname2]);
const isSameSite = a.value && b.value && a.value === b.value;
```

Happy to write test cases. Is there some kind of template / platform for that?

Agree that a "Site" API could be good, but I believe @Rob--W felt this naming might be a bit too general.

Another thing to bear in mind is that Firefox 140 is coming out in under 4 weeks. If possible, I would like to try to get this functionality included, as it is an ESR release supported for a year. So it would be great if we can try to finalize this proposal asap.

Reviewer 5: For a Swiss-army knife "Public Suffix List API" I agree with Rob: don't use the vague term "Site". I just don't really understand why an extension needs all the gory PSL details beyond that. I deeply feel the need to know what `<current browser>` thinks is a site and was proposing a simple alternative to a complex API, but if y'all need the gory details then complex API it has to be.

I'd love to expose the "site" concept to the regular DOM, too, and if it's there then extensions that didn't need the complex API could use the simple thing from that. But that could be a fingerprinting vector that allows scripts to detect exactly what browser and version of the PSL is in use. The brand and major version is already available to the site through the UA, but this could be a(nother) way to detect spoofed UAs.

> Another thing to bear in mind is that Firefox 140 is coming out in under 4 weeks

7-8 weeks until release, but yeah, effectively the deadline is < May 26 when we go to Beta.
https://whattrainisitnow.com/release/?version=140

[I don't think that's a realistic timeline to get an API standardized enough that Firefox leadership would let it go to Release, but maybe the rules are less strict for extension APIs because there are already a lot of cross-browser incompatibilities? @Rob--W would have a better perspective on that]

Reviewer: If this were to be a Firefox-only extension API, we would usually just land it and let it ride the train. In this case we'd like to have a cross-browser extension API because of the common browser-independent demand for it.

On the API design question: I would not introduce multiple similarly named methods that have the same signature but just slightly different behavior. Having methods with sane defaults and appropriately chosen options to configure the behavior seems better.

Here is an example; the property names are just for illustration, I have no strong attachments to the proposed names here.

```
browser.publicSuffix.getRegistrableDomain(
  string hostname, // from url.hostname
  { // Optional options object
    // Whether to accept domains that are unknown to the PSL
    // (This includes IP addresses)
    allowUnknown: boolean, // default false

    // Etc.
  }
)
// returns string or rejects if input is invalid
```

Lastly, if you are aiming for a quicker resolution, I recommend to focus on functionality that is really needed by extension authors. In theory it could be nice to return parsed information, but if the requested level of detail is not universally available across browsers, then it would be more difficult to have cross-browser adoption.

When I look at a random library, psl on npm for example, I see that it has a very small API surface yet 1k dependents covering 30M weekly downloads. That makes me wonder whether there is a minimal yet useful MVP that we can aim for.

Reviewer 6: > It would be better if "site" information were available in a regular DOM API, not limited to extensions. 

@dveditz For information (without hijacking too much this thread), there's been an unsuccessful attempt in the past. When this initiative started, there wasn't a lot of interest in making a second proposal for a DOM API even if your suggestion makes sense to me.

Author: @Rob--W With your proposed API, `getRegistrableDomain` can return IP addresses and hostnames that are neither registrable nor domains on the internet:

```js
getRegistrableDomain("1.2.3.4",     { allowUnknown: true }) // 1.2.3.4
getRegistrableDomain("bob.homenet", { allowUnknown: true }) // bob.homenet
getRegistrableDomain("homenet",     { allowUnknown: true }) // homenet or null?
```

To avoid this semantic issue, I suggested we could have a `get()` method that strictly returns registrable domains, and a `parse()` method for everything else:

```js
registrableDomain.get("1.2.3.4")     // null
registrableDomain.get("bob.homenet") // null
registrableDomain.get("homenet")     // null

registrableDomain.parse("1.2.3.4")     // {value: "1.2.3.4",     type: "IPAddress"}
registrableDomain.parse("bob.homenet") // {value: "bob.homenet", type: "UnknownSuffix"}
registrableDomain.parse("homenet")     // {value: "homenet",     type: "NoSuffix"}
```

I thought simple, short method names like `get` and `getAll` would be more similar to existing APIs such as the [cookies](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/cookies) API.

That said, if people think the `getRegistrableDomain()` API naming is ok (i.e. we allow a bit of flexibility over what constitutes a "registrable domain"), then I'm totally happy to go with it.

I do also agree with @dveditz and @simon-friedberger that having a simple way of knowing "what is the current site?" is useful. Are we ok to call this a "registrable domain" (noting that it includes IP addresses and intranet hostnames that might not have a suffix)?

Author: @Rob--W The [psl library](https://www.npmjs.com/package/psl) has the exact same methods I'm proposing, i.e.`get()` and `parse()`

Reviewer: > To avoid this semantic issue, I suggested we could have a `get()` method that strictly returns registrable domains,

I am supportive of a method to retrieve the strictly registered domain for an input (and return null otherwise, unless the input is really malformed, such as being a number). The extension can handle fallbacks themselves.

I mentioned options (similar to "strict" that was suggested before) because I would consider that to be preferable to copy-pasting API namespaces that are nearly identical with minor differences. I am okay with being strict and letting extensions handle non-PSL functionality themselves.

> and a `parse()` method for everything else:

What use case is fulfilled by this proposal that cannot be covered by the get method?

> I thought simple, short method names like `get` and `getAll` would be more similar to existing APIs such as the [cookies](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/cookies) API.
> 
> That said, if people think the `getRegistrableDomain()` API naming is ok (i.e. we allow a bit of flexibility over what constitutes a "registrable domain"), then I'm totally happy to go with it.

`browser.publicSuffix.registrableDomains.get` works, but if there are only a few methods, I would prefer a single namespace, in which case the name should reflect its purpose, `browser.publicSuffix.getRegistrableDomain`.

Author: > > and a `parse()` method for everything else:
> 
> What use case is fulfilled by this proposal that cannot be covered by the get method?

The use cases described by @dveditz and @simon-friedberger that are based on determining a hostname's "site", which may be:

1. a registrable domain
2. an IP address
3. an intranet hostname with a non-public (i.e. unknown) suffix, or no suffix

Author: @Rob--W @simon-friedberger @dveditz I have updated the proposal b874592521b2ce45662554295f58f9c784a45efa as follows:

* Made reference to the use cases based on "site" as opposed to "registrable domain"
* Added a `strict = false` option to allow including IP addresses and unknown TLDs
* Added a `parse()` method for more fine-grained control

I have kept the method names as they were previously, i.e. `getRegistrableDomain()` etc.

| foo.xn--mgbx4cd0ab | this is an IDN, but Punycode                     | foo.مليسيا             |
| foo.xn--mgbx4cd0ab | as above, with `punycode = true`                 | foo.xn--mgbx4cd0ab     |
|                    | empty string                                     | Error                  |
| .                  | no domain labels                                 | Error                  |
| example..com       | contains an empty domain label                   | Error                  |

#### 8. Batching

For [Use Case #2](#2-group-domains-in-ui), function `getRegistrableDomains()` enables
extensions to get multiple registrable domains with a single API call.

The returned promises's resolved array should contain a `RegistrableDomains` object
for each item in the input `domains` parameter, with the same ordering.

An error during the lookup of any of the domain names does not cause the returned promise
to be rejected. Instead, each such error is stored in the `error` key of the corresponding `RegistrableDomains` object for the domain name in question.

### New Permissions

| Permission Added | Suggested Warning |
| ---------------- | ----------------- |

> | publicSuffix     | N/A               |

Reviewer 3: This seems fine but it should explain why the browser is not more fingerprintable, for example, by arguing that the browser version is already available.

Author: I will update the document as you suggest.

Author: Addressed in c2171564bebfbd15314e2e147a613129044a4753: noted the fingerprintability

### Manifest File Changes

There are no changes to the manifest.

## Security and Privacy

### Exposed Sensitive Data

The only data exposed by this API is the [public suffix list](https://publicsuffix.org/list/).

### Abuse Mitigations

> This does not expose any new non-public data so there are no new abuse vectors.

Reviewer 3: As stated above, it does expose which version of the PSL is used.

Author: I will note this in the document.

Author: Addressed in c2171564bebfbd15314e2e147a613129044a4753: noted the fingerprintability

### Additional Security Considerations

N/A

## Alternatives

### Existing Workarounds

Developers can download the PSL dataset, bundle it with their extensions, and implement
logic that parses and interprets the dataset in order to determine the
*registrable domain / eTLD+1* for a domain name. There are several drawbacks to
this approach:

1. Potential for inconsistencies in the determination of *registrable domain / eTLD+1*
by the extension and the host browser, due to differences in the version
of the PSL dataset used, and differences in the implementations that the host
browser and the extension use in order to interpret this dataset.
2. Increased complexity and maintenance costs of extensions.
3. Increased performance overhead of extensions due to bundling of bulky
PSL dataset, leading to increase in memory, disk usage and increase in CPU usage
due to possibly suboptimal extension implementations and repeating work already
done by the host browser.

### Open Web API

> The purpose of this API is to eliminate the potential for inconsistency between

Reviewer 3: This is a bit at odds with the desire to retrieve version information.

Author: Noted, however the browser's PSL version may be of use to extensions.

##### 4.1 Recommendation

It is recommended that registrable domains should be returned as Punycode by default,
but the API should also provide an option to convert these to Unicode.

#### Use Cases

#### 1. Filter Requests by Organization

This API is relevant to any extension that does both of the following:

* acts on web requests automatically while the user browses
* allows the user to choose which domain names the extension should act on,
and which should be ignored by the extension

In effect, such extensions allow users to create request filtering rules that
restrict the extension to acting on only web requests having specific domain names.

When choosing domain names for the request filter, users may want to be able to choose
all domain names for an organization without having to enter them each in individually.
The user may want to specify request filtering rules as follows:

| # | Filtering rule type | User value | Examples of domains affected |
|:-:|---------------------|:----------:|-----------------------------:|
| 1 | organizational base domain and all possible subdomains | myorg.co.uk | myorg.co.uk<br>subdomain1.myorg.co.uk<br>subdomain2.myorg.co.uk |

> | 2 | organization name and all possible eTLDs | myorg | myorg.co.uk<br>myorg.com<br>myorg.net |

Reviewer: The two items here are distinct: "get base domain for full domain" vs "list all eTLD+1 for every TLD" cases are very distinct. The first case is obvious, the second seems niche to me. Section 1 is referenced in many places, including "Known Consumers", and it is not immediately obvious which of the two filtering rules you have in mind. I recommend to split the scenarios, and declare the second as out-of-scope.

Is the `organization name and all possible eTLDs` scenario covered by the proposed API? My first thought is that to support it, the API would have to list all known eTLDs. Another interpretation is for the method to be used to validate that what the extension thinks is a TLD, is really a TLD.

One of the listed "known consumers" is Violentmonkey, and I guess that it is for the "magic wildcard" ".tld" suffix from user scripts. The API does not offer a way to enumerate all TLDs, so it is not obvious how one would construct `organization name and all possible eTLDs` through the API. Violentmonkey and Greasemonkey have historically supported `tld` by bundling or fetching the TLD list (https://github.com/violentmonkey/violentmonkey/issues/214).

Note: historically, Google has used many TLDs, so a `.tld` wildcard domain could have been the easiest way (from the user script developer perspective) to match "all Google domains" with a single pattern. In practice, exploding this to ALL of TLDs on the ever-growing list of TLDs is not ideal. Not all eTLD+1 are real Google domains. For example, here is how I create the list of Google domains in one of my own extensions: https://github.com/Rob--W/dont-track-me-google/blob/master/tools/update-manifest.js + [manifest.json example](https://github.com/Rob--W/dont-track-me-google/blob/a73ab8b3bfa9794c75cb4e111c6424e31e6c019e/manifest.json#L15-L208).

In theory, expanding `*.tld` to a list of match patterns could match everything as desired, but in practice it would just result in an explosion of match patterns where most are not relevant.

Author: > the API would have to list all known eTLDs

My expectation would be that you would specify a single rule such as `some-organization.*` or `some-organization.tld`, and then whenever the browser subsequently made a web request you would do the following:

1. get the *first label only* of the request's registrable domain
2. if that label equals `some-organization`, you have a match

So in both of the "Filtering rule types" in the table, you are doing the same thing: getting the registrable domain from a web request. The difference is what you then do with that registrable domain in order to determine if there is a filtering-rule match or otherwise.

Reviewer: Your explanation makes sense for matching web requests at runtime, but it would not work for declarative (user) script cases.

For declarative cases, I would consider that explicitly out of scope for the current proposal and defer to https://github.com/w3c/webextensions/issues/394

For your scenario, matching wildcard tlds at runtime, I expect that an extension could use the API like this:

```
function matchesTLDwildcard(pattern, host) {
  // E.g.pattern = "example", parsed from example.tld
  const suffix = browser.publicSuffix.getKnownPublicSuffix(host);
  if (!suffix) {
    return false;
  }
  const patternWithSuffix = pattern + "." + suffix;
  return patternWithSuffix === host;
```

Author: Agreed that declarative use cases are out of scope.

Your example method looks good. It could also be implemented with `getRegistrableDomain()` as follows:

```js
function matchesTLDwildcard(pattern, host) {
  // E.g.pattern = "example", parsed from example.tld
  const registrableDomain = browser.publicSuffix.getRegistrableDomain(host);
  if (!registrableDomain) {
    return false;
  }
  return registrableDomain.startsWith(pattern + ".");
}
```

Author: Addressed in 5ba391fd8441e6754317e74216dad919e4ec9021 - noted that the declarative use case is out-of-scope

extension to allow users to create containers that isolate organizational base domains:
[PR #2352](https://github.com/mozilla/multi-account-containers/pull/2352).
Currently, users must manually curate their containers to capture all anticipated subdomains
of an organization's base domain, and this manual approach is cumbersome and error-prone.

In addition to this, an analysis has been done of the relevant use cases of a selection of
other extensions that currently roll their own solutions in order to parse the PSL, and the
use cases are set out in the [Use Cases](#use-cases) section. Although the extension
authors have not been consulted directly, it is assumed that they would stand to benefit
from this API due to the reasons stated in the [Objective](#objective) section.

| Extension | Relevant use cases | Description |
|-----------|--------------------|-------------|
| [multi-account-containers](https://addons.mozilla.org/en-US/firefox/addon/multi-account-containers/) | #1, #2 | isolate requests into containers by organizational base domain, view all domains grouped by organization |

> | [uBlock Origin](https://addons.mozilla.org/en-US/firefox/addon/ublock-origin/) | #1, #2, #3 | create organization/3rd-party request-blocking rules, view request blocking status by organization, create blocking rules for specifically 3rd-party requests |

Reviewer: FYI uBlock Origin uses its own library at https://github.com/gorhill/publicsuffixlist.js . Although it defines a richer API (https://github.com/gorhill/publicsuffixlist.js/blob/master/README.md#api), the extension only uses `publicSuffix.getDomain`.

Author: Similarly to uBlock Origin, the proposal's API is primarily based around `getRegistrableDomain()` - see latest commit 5ba391fd8441e6754317e74216dad919e4ec9021

| [adblock-plus](https://addons.mozilla.org/en-US/firefox/addon/adblock-plus/) | #1, #3 | create organization/3rd-party request-blocking rules |
| [duckduckgo-for-firefox](https://addons.mozilla.org/en-US/firefox/addon/duckduckgo-for-firefox/) | #3 | exclude requests from tracking detection unless they are 3rd-party |
| [violentmonkey](https://addons.mozilla.org/en-US/firefox/addon/violentmonkey/) | #1 | run scripts automatically for requests having a specific organization name plus any eTLD |

## Specification

### Schema

A new API `publicSuffix` is added as follows:

```ts
namespace publicSuffix {

  // METHODS

  // Determines if the given hostnames have the same registrable domain.
  export function hasSameRegistrableDomain(
    hostname1: string,
    hostname2: string,
    options?: RegistrableDomainOptions,
  ) : Promise<boolean>;

  // Gets the registrable domain of a given hostname.
  export function getRegistrableDomain(
    hostname: string,
    options?: RegistrableDomainOptions,
  )

>   : Promise<string | null>;

Reviewer: I'm wondering whether it is feasible to make this API synchronous. We usually require new extension APIs to be asynchronous unless a good reason is given otherwise.

In practice, the example extensions I checked in your list currently use a library with a synchronous `getDomain` method, and generally rewriting already-sync code to use an asynchronous method is difficult. Moreover, the bulk `getRegistrableDomain` method proposed here shows that the basic `getRegistrableDomain` method already appears to have too much overhead that necessitates a bulk query method.

In Firefox, the internal [effectiveTLDService](https://searchfox.org/mozilla-central/rev/4bd620c218a01568d541453ce3b215ebf55bca63/netwerk/dns/nsIEffectiveTLDService.idl) API can be used from the parent and content process, including extensions. There is no implementation constraint for requiring this to be implemented in the parent process.

In Chromium, I see at least one use of `registry_controlled_domains` in the renderer ([`LocalDOMWindow::IsCrossSiteSubframe` in `local_dom_window.cc`](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/frame/local_dom_window.cc;l=366;drc=10109fb2e0f222a28c5cbbe24dccd784a0ad7e6d)), which suggests that the information may be available in the child process too.

In WebKit, I see `topPrivatelyControlledDomain` being used [`areSameSiteIgnoringPublicSuffix` in `Document.cpp`]( https://searchfox.org/wubkat/rev/9db84d0aa73958ea6a2639a69849bd734c241352/Source/WebCore/dom/Document.cpp#4500-4502), suggesting that the information may be available in the content process as well.

Since the current implementations offer the functionality in the child process, and the more ergonomic version of the API is for it to be synchronous, I'm favoring the APIs to be non-async.

@oliverdunk @xeenon Thoughts?

Author: I have gone ahead and changed this proposal's API to sync in commit 5ba391fd8441e6754317e74216dad919e4ec9021 However, I can revert this to async if necessary.

Reviewer 2: This sounds reasonable to me, however I would ultimately defer to @rdcronin here.

  // Gets the registrable domain of each hostname in a given list of hostnames.
  export function getRegistrableDomains(

>     hostnames: Iterable<string>,

Reviewer: ```suggestion
    hostnames: Array<string>,
```

Make it a plain array (also below). An iterable implies that one could pass an Iterable, but in practice I think that most if not all implementations accept plain arrays only.

I'm suggesting `Array<string>` because you uses that syntax below, but `string[]` would have been equally fine. As long as the intention is clear the exact syntax is not super important.

Author: The return value of this method is modelled on [Promise.all()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all), which accepts an iterable rather than an array as its input parameter.

Author: Addressed in 5ba391fd8441e6754317e74216dad919e4ec9021 - this batching method has been removed in favour of a synchronous API

    options?: RegistrableDomainOptions,
  )
  : Promise<Array<RegistrableDomainResult>>;

  // Determines which one of the following kinds of value applies to each hostname
  // in a list of hostnames:
  //   RegistrableDomain
  //   PublicSuffix
  //   IPAddress
  //   Unknown
  //   Invalid
  export function parse(
    hostnames: Iterable<string>,
  )
  : Promise<Array<ParseResult>>

  // Gets the value of the VERSION metadata field in the PSL dataset if available
  export function getVersion(): string | null;

  // INTERFACES

  // Options that may be passed to the API's methods to control their behaviour.
  interface RegistrableDomainOptions {
    // If true, each resulting registrable domain should be encoded as Unicode.
    // Default = false (Punycode)
    unicode?: boolean,
    // If false, IP addresses and hostnames lacking a known eTLD are

>     // treated as having registrable domains.

Reviewer: ```suggestion
    // treated as being registrable domains.
```

Author: I had originally written "being/having" but changed it to "having" for brevity!

In the case of IP addresses, "being" is correct, because the IP address *is* the registrable domain.

In the case of hostnames lacking a known public suffix, the registrable domain is the *final label* only, which may or may not be different to the input hostname - so "being/having" would be correct.

Author: Addressed in 5ba391fd8441e6754317e74216dad919e4ec9021

#### 6. Invalid hostname

The promises returned by this API's methods should reject with an error if a hostname
passed as an input parameter meets any of the following criteria:

* Contains a character that is invalid in an Internationalized Domain Name (IDN) - e.g. symbols, whitespace
* Is an empty string
* Is equal to `'.'`
* Contains empty domain labels (i.e. any occurrences of `'..'`)

#### 7. Summary of behaviours

The following table sets out the eventual settled state of the promise returned by
`getRegistrableDomain()` for different classes of input `hostname` parameter:

| Input hostname     | Description                                      | Registrable domain     |
|:-------------------|:-------------------------------------------------|-----------------------:|
| example.net        | eTLD+1                                           | example.net            |
| www.example.net    | eTLD+2                                           | example.net            |
| net                | is an eTLD itself, single-label                  | null                   |
| net                | as above, with `strict === false`                | net                    |

> | github.io          | is an eTLD itself, multi-label                   | github.io              |

Reviewer: By which rule/criterion does the `github.io` input result in the `github.io` output?

The proposal here states that the longest public suffix is used.
The URL spec has the following step in the algorithm to determine a "registrable domain ( https://url.spec.whatwg.org/#host-registrable-domain ):

> If host’s [public suffix](https://url.spec.whatwg.org/#host-public-suffix) is null or host’s [public suffix](https://url.spec.whatwg.org/#host-public-suffix) [equals](https://url.spec.whatwg.org/#concept-host-equals) host, then return null.

By that logic, with `github.io`'s (longest) public suffix being `github.io`, then github.io is an invalid registrable domain. However, one can actually visit https://github.io/

Author: My understanding is that a strict registrable domain (eTLD+1) should have a minimum of 2 labels, i.e. `[+1].[eTLD]`

Therefore I would expect the algorithm to run as follows:

1. Ignore the leading label (this may or may not be the `+1`)
2. Get the public suffix of the remainder, which may itself be a public suffix (this is the `eTLD`)
3. Prepend the preceding label (the `+1`) to the obtained public suffix

So for `github.io` (which is itself a public suffix, as is `io`), these steps would yield the following:

| Input hostname | +1 | eTLD |
|-|-|-|
| github.io | github | io |

The algorithm in the URL spec seems a little unclear, because it states the following:

> return null or a [domain](https://url.spec.whatwg.org/#concept-domain) formed by host’s [public suffix](https://url.spec.whatwg.org/#host-public-suffix) and the [domain label](https://url.spec.whatwg.org/#domain-label) preceding it, if any.
> [...]
> If host’s [public suffix](https://url.spec.whatwg.org/#host-public-suffix) is null or host’s [public suffix](https://url.spec.whatwg.org/#host-public-suffix) [equals](https://url.spec.whatwg.org/#concept-host-equals) host, then return null

Unless I'm reading it wrong, it seems that the first point allows returning a public suffix without a preceding domain label (note the "if any" wording), whereas the second point suggests this is not possible (the result must be null).

Reviewer: A reference matching algorithm is documented at https://github.com/publicsuffix/list/wiki/Format#formal-algorithm

If you want to explore a way to get the eTLD+1 for a given input, try;

`Services.eTLD.getBaseDomainFromHost("github.io")`

throws `NS_ERROR_INSUFFICIENT_DOMAIN_LEVELS`

You can find an example of the expected result in these unit tests:
- `kyoto.jp` and `jp` are both on the PSL, and checked at: https://searchfox.org/mozilla-central/rev/9305025f453b8eedbd0bd38c87c54be73e3cd064/netwerk/test/unit/data/test_psl.txt#52-53
- test helper that calls getBaseDomainFromHost: https://searchfox.org/mozilla-central/rev/9305025f453b8eedbd0bd38c87c54be73e3cd064/netwerk/test/unit/test_psl.js#17-19,22,38

Author: Given some public suffixes can themselves have websites - examples being [github.io](github.io) and [blogspot.com](blogspot.com) as you note - would it be an idea to clarify/improve the registrable domain algorithm implementation in the host browser itself, instead of introducing a workaround (i.e. `allowPlainSuffix`) in this API?

Author: Addressed in 5ba391fd8441e6754317e74216dad919e4ec9021 - noted that by default known eTLDs cannot have/be registrable domains

`strict = false` option, in that it gets a hostname's registrable domain if possible,
or otherwise the most appropriate alternative value.

The key difference is that the `parse()` method also returns an indicator of the
kind of value that was returned, not just the value itself. In addition,
the `parse()` method does not throw an error if the hostname is invalid.

The `kind` value of the returned object must be one of the following:
`RegistrableDomain`, `PublicSuffix`, `Unknown`, `IPAddress`, `Invalid`.

##### Examples: parse result

| Input hostname  | Parse: value   | Parse: kind       | Explanation                       |
|-----------------|----------------|-------------------|-----------------------------------|
| foo.example.com |    example.com | RegistrableDomain | eTLD+1 with a known eTLD          |
| com             |    com         | PublicSuffix      | Known eTLD lacking a +1 label     |
| printer.homenet |    homenet     | Unknown           | Lacking a known eTLD              |
| 1.2.3.4         |    1.2.3.4     | IPAddress         | An IP address                     |
| *.com           |    null        | Invalid           | Contains invalid character '*'    |

> ##### 9.1 Justification for parsing

Reviewer: This `parse` part of the API came from the part of the discussion around https://github.com/w3c/webextensions/pull/676/#discussion_r2068145782.

I'm having a hard time trying to come up with a scenario where an extension would use these methods. If I compare it with the list of Use Cases in the document, I don't see an obvious answer either. That's why later in the thread [I asked](https://github.com/w3c/webextensions/pull/676/#discussion_r2072421232) for the use case of `parse`, and you [responded](https://github.com/w3c/webextensions/pull/676/#discussion_r2072495819):

> The use cases described by @dveditz and @simon-friedberger that are based on determining a hostname's "site", which may be:
> 
> 1. a registrable domain
> 2. an IP address
> 3. an intranet hostname with a non-public (i.e. unknown) suffix, or no suffix

These distinctions on their own do not immediately warrant a dedicated method in my eyes:

- case 1 is covered by `getRegistrableDomain` / `getRegistrableDomains`.
- case 2 can be covered in the extension with relatively simple logic and no new extension APIs.
- case 3: any domain can potentially be an intranet hostname. The specific example that Simon gave was `home.arpa` and `localhost`. Your [response](https://github.com/w3c/webextensions/pull/676/#discussion_r1974546953) to the `home.arpa` example is "The API returns null if a string is a domain with an unknown suffix.". Atlhough `home.arpa` is a public, so is `arpa`, so I would expect `home.arpa` to be returned by getRegistrableDomain, even with strict=true. This would be comparable with the `github.io` example in your table.

Re https://github.com/w3c/webextensions/pull/676#discussion_r2072262652

> @Rob--W The [psl library](https://www.npmjs.com/package/psl) has the exact same methods I'm proposing, i.e.`get()` and `parse()`

If I compare it with the [`parse` method from the psl library on npm](https://www.npmjs.com/package/psl#api) for example, I note that although the method name is the same, the outputs differ significantly. Your proposal returns a value + "kind", whereas the npm library returns tld, sld, domain, subdomain for a given input (sld = eTLD+1).

Author: I introduced the `parse()` method mainly because I was unsure about returning an IP address from a method called `getRegistrableDomain()`, because an IP address doesn't seem to be a registrable domain in the strictest sense. By contrast, returning an IP address from `parse()`, a more neutrally-named and general-purpose method, seems fine.

Also, reviewers were previously proposing either `allowUnknown` or `strict` as options, neither of which would allow distinguishing between *both* of the following cases:

1. domain name vs IP address
2. known vs unknown suffix

It seems people are ok with returning a non-registrable domain (e.g. an IP address) from a method called `getRegistrableDomain()`. Also, with the proposal to split option `allowUnknown/strict` into `allowIP` and `allowUnknownSuffix`, we provide the flexibility to handle both of the cases above.

It is still the case that you will not know whether the result of `getRegistrableDomain()` is an IP address or a domain name, nor whether the suffix is known or unknown, if you call it with options `{ allowIP: true, allowUnknownSuffix: true }`.

In theory, given a hostname, you could infer what kind of "registrable domain approximation" it is or has by calling `getRegistrableDomain()` multiple times with different options, but that would be at the expense of multiple async calls, which would be costly performance-wise. However, if we move to a synchronous API, this performance issue largely goes away.

I can also see the motivation to keep the API as concise as possible by keeping the number of methods to a minimum.

Re. the `home.arpa` example, apologies I misunderstood what Simon was saying. I agree that since `arpa` and `home.arpa` are public suffixes, the registrable domains would be:

| Input hostname | Registrable domain | Matched Suffix |
|-:|-:|-:|
| home.arpa | home.arpa | arpa |
| my-device-hostname.home.arpa | my-device-hostname.home.arpa | home.arpa |

Author: Addressed in 5ba391fd8441e6754317e74216dad919e4ec9021 - removed the `parse()` method, noted the concerns that had led to its inclusion, added the additional options to `getRegistrableDomain()`

which may be an IP address or a domain name.

An example of such a use case is Firefox's [Search vs Navigate](#4-search-vs-navigate),
which involves determining if an entry in the URL bar is a navigable site,
or a search term. If this functionality was based purely on the return value
of `getRegistrableDomain()`, i.e. navigate if nonnull or search if null,
then IP addresses would incorrectly cause a search. By using the `allowIP` option,
the return value for an input IP address would be the IP address itself instead of null,
thereby causing the desired result of navigating instead of searching.

Option `allowPlainSuffix` only exists because there are domains that do not have
a registrable domain, due to themselves being PSL eTLDs, but can still be
navigated to, such as github.io and blogspot.com.

##### 2.2.3 Options: Discussion

The effect of the options is that `getRegistrableDomain()` may return values
that are not registrable domains in the strictest sense, e.g. they may
be IP addresses.

The author of this proposal is of the view that:

1. Any method named `getXYZ()` should return a value of type `XYZ`. Therefore

> `getRegistrableDomain()` may not be the most suitable name, since it does

Reviewer: I'm fine with either name - `publicSuffix.getRegistrableDomain()` or `publicSuffix.getDomain()`.

The current descriptions of the API schema frames the options as ways to change "registrable domain". I'd view it differently: independently of the method name, the definition of "registrable domain" is unchanged. The default return value is a registrable domain (as defined by URL spec). What the options do is to alter the return value to include more than just a registrable domain.

For clarity, it may make sense to change "Registrable domain" to "Returned domain" in the comparisons/tables, so that even if someone did not read the whole document, that they would not be confused when they see examples under such a column heading.

Author: Addressed in f1d81e5c10e93748e49c7a018ac9351b404fdd6d - changed the method names to the below as discussed, and reworded "registrable domain" to "returned domain" where appropriate:

```js
publicSuffix.getDomain()
publicSuffix.getKnownSuffix()
publicSuffix.isKnownSuffix()
```

N/A

## Alternatives

### Existing Workarounds

Developers can download the PSL dataset, bundle it with their extensions, and implement
logic that parses and interprets the dataset in order to determine the
registrable domain for a domain name. There are several drawbacks to
this approach:

1. Potential for inconsistencies in the determination of registrable domains
by the extension and the host browser, due to differences in the version
of the PSL dataset used, and differences in the implementations that the host
browser and the extension use in order to interpret this dataset.
2. Increased complexity and maintenance costs of extensions.
3. Increased performance overhead of extensions due to bundling of bulky
PSL dataset, leading to increase in memory, disk usage and increase in CPU usage
due to possibly suboptimal extension implementations and repeating work already
done by the host browser.

### Open Web API

The purpose of this API is to eliminate the potential for inconsistency between
the host browser and its hosted extensions. The simplest way of achieving this
is for extensions to access this functionality via the host browser itself rather

> than via some external source, such as an Open Web API.

Reviewer 2: By "Open Web API", we mean any API implemented by the browser and exposed in JavaScript. These can absolutely use internal sources of data, so I don't think this is a compelling argument. I would perhaps update this to focus more on the fact that:

- There is a higher bar for introducing web APIs, and we haven't seen much interest in moving forward a proposal like this in the past. Therefore, we would prefer to start with extensions, and can always propose a web API later if this work proves useful and there is appetite.
- The PSL is not appropriate for use in all circumstances. Extensions have a very compelling set of use cases that match browser use cases, but there is not a universal agreement this is the case more generally.

Author: Addressed in f1d81e5c10e93748e49c7a018ac9351b404fdd6d - I have updated this section as you suggest

Reviewer 2: Thanks!

| Extension | Relevant use cases | Description |
|-----------|--------------------|-------------|
| [multi-account-containers](https://addons.mozilla.org/en-US/firefox/addon/multi-account-containers/) | #1, #2 | isolate requests into containers by organizational base domain, view all domains grouped by organization |
| [uBlock Origin](https://addons.mozilla.org/en-US/firefox/addon/ublock-origin/) | #1, #2, #3 | create organization/3rd-party request-blocking rules, view request blocking status by organization, create blocking rules for specifically 3rd-party requests |
| [adblock-plus](https://addons.mozilla.org/en-US/firefox/addon/adblock-plus/) | #1, #3 | create organization/3rd-party request-blocking rules |
| [duckduckgo-for-firefox](https://addons.mozilla.org/en-US/firefox/addon/duckduckgo-for-firefox/) | #3 | exclude requests from tracking detection unless they are 3rd-party |
| [violentmonkey](https://addons.mozilla.org/en-US/firefox/addon/violentmonkey/) | #1 | run scripts automatically for requests having a specific organization name plus any eTLD |

## Specification

### Schema

A new API `publicSuffix` is added as follows:

```ts
namespace publicSuffix {

  // METHODS

>   // Determines if the given hostname is itself a known eTLD (i.e. in the PSL).
>   export function isKnownSuffix(
>     hostname: string,

Reviewer: ```suggestion
  // Determines if the given suffix is itself a known eTLD (i.e. in the PSL).
  export function isKnownSuffix(
    suffix: string,
```

Nit: The input is not a hostname but a suffix.

Author: I am using the term "hostname" in the sense of the [hostname property](https://developer.mozilla.org/en-US/docs/Web/API/URL/hostname) of javascript's URL class:

```js
const hostname = new URL("https://co.uk").hostname;
const isKnownSuffix = publicSuffix.isKnownSuffix(hostname);
```

The input parameter may be any hostname, and if it is anything other than a known suffix (e.g. an IP address, or an eTLD+1), then the method returns `false`.

In section "Behaviours: 4. Invalid hostname" I state:

> This API's methods should throw an error if a hostname passed as an input parameter
> [contains invalid characters]

I intended that same error-throwing behaviour to apply to this method too, so referring to the input parameter as a "hostname" is a way of conveying that intent.

  )
  : boolean;

  // Gets the known eTLD, if any, of a given hostname.
  export function getKnownSuffix(
    hostname: string,
  )
  : string | null;

  // Gets the registrable domain of a given hostname.
  export function getDomain(
    hostname: string,
    options?: DomainOptions,
  )
  : string | null;

  // Gets the value of the VERSION metadata field in the PSL dataset if available
  export function getVersion(): string | null;

  // INTERFACES

  // Options that may be passed to the API method to control its behaviour.
  interface DomainOptions {
    // If true, the returned domain should be encoded as Unicode.
    // Default = false (Punycode)
    unicode?: boolean,

>     // If true, the returned domain may be an IP address.

Reviewer: ```suggestion
    // If true, the input hostname may be an IP address, and be returned as is.
```

Author: Addressed in e0f283531330fb48febfad686795793a41dc630f

    // Default = false
    allowIP?: boolean,

>     // If true, the returned domain may be a known eTLD.

Reviewer: ```suggestion
    // If true, the input hostname may be a known eTLD without preceding label, and be returned as is.
```

Author: Addressed in e0f283531330fb48febfad686795793a41dc630f

    // Default = false
    allowPlainSuffix?: boolean,

>     // If true, the returned domain may lack a known eTLD.

Reviewer: ```suggestion
    // If true, the input hostname may lack a known eTLD, and the returned domain consists of the penultimate two domain labels of the input.
```

Author: Addressed in e0f283531330fb48febfad686795793a41dc630f

##### Example of Determining if Requests are Third-Party

| Req. | Domain                      | Registrable Domain          | Third-party? |
|:----:|----------------------------:|----------------------------:|:------------:|
| #1   | foo.com                     | foo.com                     |              |
| #2   | bar.com                     | bar.com                     | Yes          |
|      |                             |                             |              |
| #1   | foo.amazonaws.com           | amazonaws.com               |              |
| #2   | bar.amazonaws.com           | amazonaws.com               | No           |
|      |                             |                             |              |
| #1   | foo.amazonaws.com           | amazonaws.com               |              |
| #2   | bar.us-east-1.amazonaws.com | bar.us-east-1.amazonaws.com | Yes          |

#### 4. Search vs Navigate

Firefox makes use of the PSL in order to determine whether to issue a search query
or whether to try a navigation, when a user enters a domain-like string in the
url bar. In such instances, a PSL lookup is made and:

* If the domain has a known eTLD, attempt to navigate.

> * If the domain has an unknown eTLD, use a search engine.

Reviewer 7: This is a use case for the PSL, but I'm not sure it's as compelling a use case for extensions themselves (since it boils down to "handle things with TLDs differently", which is already described)

Author: This use case was provided by a reviewer during review of this proposal to justify the inclusion of functionality to distinguish between known vs unknown suffixes. I will add an update to note that this specific use case may not apply directly to extensions.

Reviewer 3: @rdcronin IMHO there are two ways the PSL is used:

1. Figuring out a site, for grouping or associating data with a site, or deleting site history, etc.
2. Figuring out if something looks like a valid domain - that's what this is and I don't think it's covered above...? Actually, it looks like all the other examples are examples of (1.) and this is the only example of (2.).

Author: Addressed in aafb544a124d7f344a4bdd810caef0e3ada6a6e6

| [violentmonkey](https://addons.mozilla.org/en-US/firefox/addon/violentmonkey/) | #1 | run scripts automatically for requests having a specific organization name plus any eTLD |

## Specification

### Schema

A new API `publicSuffix` is added as follows:

```ts
namespace publicSuffix {

  // METHODS

  // Determines if the given hostname is itself a known eTLD (i.e. in the PSL).
  export function isKnownSuffix(
    hostname: string,
  )
  : boolean;

  // Gets the known eTLD, if any, of a given hostname.
  export function getKnownSuffix(
    hostname: string,
  )
  : string | null;

  // Gets the registrable domain of a given hostname.
  export function getDomain(
    hostname: string,
    options?: DomainOptions,
  )
  : string | null;

  // Gets the value of the VERSION metadata field in the PSL dataset if available
  export function getVersion(): string | null;

  // INTERFACES

  // Options that may be passed to the API method to control its behaviour.
  interface DomainOptions {

    // If true, the returned domain should be encoded as Unicode.
    // Default = false (Punycode)

>     unicode?: boolean,

Reviewer 7: Could this be an enum instead, to allow future encoding options?

Author: Yes, I will rename this option from `unicode` to `encoding`, as follows:

```ts
interface DomainOptions {

  // Determines how the returned domain should be encoded.
  // Default = punycode
  encoding?: DomainEncoding,

  // etc.
}

enum DomainEncoding {
  punycode,
  unicode,
}
```

Author: Addressed in aafb544a124d7f344a4bdd810caef0e3ada6a6e6

    // If true, and the input hostname is an IP address, then this is returned as-is.
    // Default = false

>     allowIP?: boolean,

Reviewer 7: nit: I'd lean towards "allowIpAddress[es]"

Author: Ok, I will rename this to `allowIpAddress`.

The other options are singular (e.g. `allowUnknownSuffix`) rather than plural, so I would favour the singular here too.

Reviewer 8: Should be `allowIPAddress`.

Author: Addressed in aafb544a124d7f344a4bdd810caef0e3ada6a6e6

    // If true, and the input hostname is itself a known eTLD (without a preceding label)
    // then this is returned as-is.
    // Default = false
    allowPlainSuffix?: boolean,

    // If true, and the input hostname lacks a known eTLD, and is neither itself
    // a known eTLD nor an IP address, then the returned domain consists of
    // the penultimate two domain labels of the input.
    // Default = false
    allowUnknownSuffix?: boolean,

>   }

Reviewer 7: Should we provide an option for whether to include private registries?  (I think in this proposal, called "implied suffixes"?)

Author: Are you referring to the "Private section" of the PSL dataset? If so, these private registries are always included, as explained in proposal section "Features of the PSL: 1. ICANN vs Private".

returning a `Promise<string>`. However, some use cases require getting lists of
registrable domains all in one go. In theory, this could be achieved by simply calling
`getDomain()` multiple times.

The problem with this approach is that there is overhead associated with an extension
calling an async function on the parent browser. For example, obtaining the registrable domains
of a list of 50 hostnames would involve making 50 async calls to the parent browser.
A batching method would allow the same result to be obtained with a single async call.

For this reason, batching method `getDomains()` was added to this API.
The method accepted an array of hostnames as input and returning a promise resolving to
an array of registrable domains. A quick mockup of the two approaches was built using
a simplified implementation of this proposal's API in a modified Firefox, and the
batching approach was about 2-3 times faster for 50 hostnames.

Unfortunately, while this offered a solution to the performance problem,
it added additional complexity to the API. To resolve this issue, the API
has now been changed to being synchronous, which has allowed the batching method

> to be removed, thereby making the API more ergonomic.

Reviewer 7: As we discussed in our meeting, I'd like us to expand this section with more details about this.  Sync APIs are against our typical best practices and, even if they are possible without significant extra architectural work today, they may not always be that way (over the past decade, we've seen many architectural shifts that have invalidated previously-held assumptions, such as subframes being in the same process or navigation being driven by the renderer process).

I'm not fundamentally opposed to these being sync, but I think we should have good reason beyond just making it a bit easier, since we may pay heavily for it in the future.  The best justification would be a scenario that's not solvable without a sync API (e.g., use in document_start), but AFAIK we haven't identified any of those.  Are there specific instances of things that are dramatically more complex to implement with async APIs?  Can we add those here as documentation for why we diverge from typical best practices?

Author: Understood, I will look into how to expand this section with further justification.

Reviewer 3: Is this async preference documented somewhere?

Reviewer 2: We mention it briefly here: https://github.com/w3c/webextensions/blob/main/proposals/proposal_template.md#schema

I think @Rob--W had some thoughts on how to update this section, so he might be able to help (sorry to sign you up for that Rob!).

Reviewer 9: @mckenfra looking forward to this proposal, and thanks for all your effort! has there been any progress regarding this topic since may?

Author: In terms of providing further justification for a sync API, @Rob--W has downloaded a list of Firefox extensions that use the PSL, and I have started examining the source code of the most popular extensions on the list. All of the extensions I have looked at so far use sync functions to get registrable domains:

| # | Extension | PSL Library Used | PSL Functions |
| - | - | - | - |
| 1 | [adblock-plus](https://addons.mozilla.org/en-US/firefox/addon/adblock-plus/) | None (own code) | Sync |
| 2 | [duckduckgo-for-firefox](https://addons.mozilla.org/en-US/firefox/addon/duckduckgo-for-firefox/) | [tldts](https://www.npmjs.com/package/tldts) | Sync |
| 3 | [adblock-for-firefox](https://addons.mozilla.org/en-US/firefox/addon/adblock-for-firefox/) | None (own code) | Sync |
| 4 | [adguard-adblocker](https://addons.mozilla.org/en-US/firefox/addon/adguard-adblocker/) | [tldts](https://www.npmjs.com/package/tldts) | Sync |
| 5 | [bitwarden-password-manager](https://addons.mozilla.org/en-US/firefox/addon/bitwarden-password-manager/) | [tldts](https://www.npmjs.com/package/tldts) | Sync |
| 6 | [wikibuy-for-firefox](https://addons.mozilla.org/en-US/firefox/addon/wikibuy-for-firefox/) | [better-psl](https://www.npmjs.com/package/better-psl) | Sync |
| 7 | [browsec](https://addons.mozilla.org/en-US/firefox/addon/browsec/) | None (own code) | Sync |
| 8 | [malwarebytes](https://addons.mozilla.org/en-US/firefox/addon/malwarebytes/) | [tldts](https://www.npmjs.com/package/tldts) | Sync |
| 9 | [noscript](https://addons.mozilla.org/en-US/firefox/addon/noscript/) | [nscl](https://github.com/hackademix/nscl/blob/main/common/tld.js) | Sync |
| 10 | [coupert](https://addons.mozilla.org/en-US/firefox/addon/coupert/) | [parse-domain](https://www.npmjs.com/package/parse-domain) | Sync |
| 11 | [nordvpn-proxy-extension](https://addons.mozilla.org/en-US/firefox/addon/nordvpn-proxy-extension/) | [tldts](https://www.npmjs.com/package/tldts) | Sync |
| 12 | [dashlane](https://addons.mozilla.org/en-US/firefox/addon/dashlane/) | [tldts](https://www.npmjs.com/package/tldts) | Sync |

@rdcronin What are you thoughts on this?

Reviewer 3: Given that DMARC is proposing publishing PSL DNS records and the PSL already uses _psl TXT records to validate that a domain owner wants to be on the PSL, it might be a good idea to use an async API in case lookups require a DNS request in the future.

Reviewer 3: > 4.10.  DNS Tree Walk
> [...]
>  If a valid DMARC Policy Record contains the "psd" tag set to "n"
       ("psd=n"), this is the Organizational Domain, and the selection
       process is complete.

[source](https://datatracker.ietf.org/doc/draft-ietf-dmarc-dmarcbis/)

Author: Addressed in aafb544a124d7f344a4bdd810caef0e3ada6a6e6

|-------------------------------|-------------------------------|
| com                           | blogspot.com                  |
|                               | members.linode.com            |
|                               | simplesite.com                |
|                               |                               |
| co.uk                         | blogspot.co.uk                |
|                               | wellbeingzone.co.uk           |
|                               | vm.bytemark.co.uk             |

In the case of eTLDs in the ICANN section, the "owner" of an eTLD is ICANN /
an affiliated registrar. In the case of eTLDs in the Private section, the
"owner" of an eTLD is a private organization offering a service for clients
to take ownership of subdomains underneath its own ICANN-suffixed domain.

At first glance, it might seem that knowing which section an eTLD is from
could give useful information as to whether or not it is an "official"
ICANN-designated eTLD. In practice, however, the distinction
between the two sections is somewhat arbitrary and not well-defined.
This is due to the inherent challenges in maintaining and curating the

> PSL dataset, which relies on the cooperation of registrars.

Reviewer 7: I still suspect that, even though these are not always robust, we'll see extensions desiring a differentiation.

That said, I'm fine with this for now -- we can default-include both private and public and add some way to differentiate in the future, if needed.

Author: Noted, yes this proposal default-includes all suffixes in the PSL (i.e. both ICANN-section and Private-section).

but sometimes one may be interested in knowing whether there is any other shorter eTLD
that might have matched the input hostname in theory. E.g. 'github.io' is a
public suffix itself, but could also be interpreted as a registrable domain
whose public suffix is 'io'.

###### Examples

| Input hostname |  Known Suffix |
|----------------|--------------:|
|      github.io |     github.io |
|  foo.github.io |     github.io |
| facebook.co.uk |         co.uk |
|    192.168.2.1 |          null |
|   green.banana |          null |

##### 2.2 Domain

Method `getDomain()` returns the input hostname's registrable domain,
as determined by running the PSL algorithm, otherwise `null`.

By default, this method returns `null` if the input hostname:

* lacks a known eTLD (i.e. in the PSL)
* is itself a known eTLD
* is an IP address - IPv4 or IPv6

##### 2.2.1 Options: Domain

In order to support different use cases including those that need to determine
a hostname's "site", additional options are provided, allowing a more
general-purpose interpretation of a domain to include not only registrable domains
but also IP addresses and domains with unknown (non-registrable) eTLDs.

> Options `allowIP`, `allowPlainSuffix` and `allowUnknownSuffix` each target

Reviewer 7: here and below, update to allowIPAddress

Author: Fixed in 2440fc039244aa89cfc2d5d8ddee634727500cff - all occurrences of `allowIP` changed to `allowIPAddress`
